
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>targets &#8212; alignparse 0.2.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="utils" href="alignparse.utils.html" />
    <link rel="prev" title="minimap2" href="alignparse.minimap2.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-alignparse.targets"></span><div class="section" id="targets">
<h1>targets<a class="headerlink" href="#targets" title="Permalink to this headline">¶</a></h1>
<p>Defines <a class="reference internal" href="#alignparse.targets.Targets" title="alignparse.targets.Targets"><code class="xref py py-class docutils literal notranslate"><span class="pre">Targets</span></code></a>, which holds <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> objects that define
alignment targets. Each <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> has some <a class="reference internal" href="#alignparse.targets.Feature" title="alignparse.targets.Feature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Feature</span></code></a> regions.</p>
<dl class="py class">
<dt id="alignparse.targets.Feature">
<em class="property">class </em><code class="sig-prename descclassname">alignparse.targets.</code><code class="sig-name descname">Feature</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">name</span></em>, <em class="sig-param"><span class="n">seq</span></em>, <em class="sig-param"><span class="n">start</span></em>, <em class="sig-param"><span class="n">end</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A sequence feature within a <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – Name of feature.</p></li>
<li><p><strong>seq</strong> (<em>str</em>) – Sequence of feature.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – Feature start in <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a>, using Python-like 0, … numbering.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – Feature end in <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> using Python-like 0, … numbering.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt id="alignparse.targets.Feature.name">
<code class="sig-name descname">name</code><a class="headerlink" href="#alignparse.targets.Feature.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Feature.seq">
<code class="sig-name descname">seq</code><a class="headerlink" href="#alignparse.targets.Feature.seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Sequence of feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Feature.start">
<code class="sig-name descname">start</code><a class="headerlink" href="#alignparse.targets.Feature.start" title="Permalink to this definition">¶</a></dt>
<dd><p>Feature start in <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a>, using Python-like 0, … numbering.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Feature.end">
<code class="sig-name descname">end</code><a class="headerlink" href="#alignparse.targets.Feature.end" title="Permalink to this definition">¶</a></dt>
<dd><p>Feature end in <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> using Python-like 0, … numbering.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Feature.length">
<code class="sig-name descname">length</code><a class="headerlink" href="#alignparse.targets.Feature.length" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of feature.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="alignparse.targets.Target">
<em class="property">class </em><code class="sig-prename descclassname">alignparse.targets.</code><code class="sig-name descname">Target</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">seqrecord</span></em>, <em class="sig-param"><span class="n">req_features</span><span class="o">=</span><span class="default_value">frozenset({})</span></em>, <em class="sig-param"><span class="n">opt_features</span><span class="o">=</span><span class="default_value">frozenset({})</span></em>, <em class="sig-param"><span class="n">allow_extra_features</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Target"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Target" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A single target sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqrecord</strong> (<em>Bio.SeqRecord.SeqRecord</em>) – BioPython sequence record of target. Must have <cite>seq</cite>, <cite>name</cite>,
and <cite>features</cite> attributes. Currently only handles + strand features.</p></li>
<li><p><strong>req_features</strong> (<em>set</em><em> or </em><em>other iterable</em>) – Required features in <cite>seqrecord</cite>.</p></li>
<li><p><strong>opt_features</strong> (<em>set of other iterable</em>) – Optional features in <cite>seqrecord</cite>.</p></li>
<li><p><strong>allow_extra_features</strong> (<em>bool</em>) – Can <cite>seqrecord</cite> have features not in <cite>req_features</cite> or <cite>opt_features</cite>?</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt id="alignparse.targets.Target.seq">
<code class="sig-name descname">seq</code><a class="headerlink" href="#alignparse.targets.Target.seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Full sequence of target.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Target.name">
<code class="sig-name descname">name</code><a class="headerlink" href="#alignparse.targets.Target.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of target.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Target.length">
<code class="sig-name descname">length</code><a class="headerlink" href="#alignparse.targets.Target.length" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Target.features">
<code class="sig-name descname">features</code><a class="headerlink" href="#alignparse.targets.Target.features" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all features as <a class="reference internal" href="#alignparse.targets.Feature" title="alignparse.targets.Feature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Feature</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Target.feature_names">
<code class="sig-name descname">feature_names</code><a class="headerlink" href="#alignparse.targets.Target.feature_names" title="Permalink to this definition">¶</a></dt>
<dd><p>List of names of all features.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Target.get_feature">
<code class="sig-name descname">get_feature</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Target.get_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Target.get_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Get <a class="reference internal" href="#alignparse.targets.Feature" title="alignparse.targets.Feature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Feature</span></code></a> by name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – Name of <a class="reference internal" href="#alignparse.targets.Feature" title="alignparse.targets.Feature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Feature</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns the feature, or raises <cite>ValueError</cite> if no such feature.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#alignparse.targets.Feature" title="alignparse.targets.Feature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Feature</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Target.get_name">
<em class="property">classmethod </em><code class="sig-name descname">get_name</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">seqrecord</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Target.get_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Target.get_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Get name of target from sequence record.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>seqrecord</strong> (<em>Bio.SeqRecord.SeqRecord</em>) – Sequence record as passed to <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Name parsed from <cite>seqrecord</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Target.has_feature">
<code class="sig-name descname">has_feature</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Target.has_feature"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Target.has_feature" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a feature is defined for this target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – Name of <a class="reference internal" href="#alignparse.targets.Feature" title="alignparse.targets.Feature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Feature</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><cite>True</cite> if target has feature of this name, <cite>False</cite> otherwise.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Target.image">
<code class="sig-name descname">image</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">color_map</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">feature_labels</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">plots_indexing</span><span class="o">=</span><span class="default_value">'genbank'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Target.image"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Target.image" title="Permalink to this definition">¶</a></dt>
<dd><p>Get image of the target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>color_map</strong> (<em>None</em><em> or </em><em>dict</em>) – To specify colors for each feature, provide a dict mapping
feature names to colors. Otherwise automatically chosen.</p></li>
<li><p><strong>feature_labels</strong> (<em>None</em><em> or </em><em>dict</em>) – Map feature names to text labels shown on plot. Otherwise
features just labeled by name.</p></li>
<li><p><strong>plots_indexing</strong> (<em>{'biopython'</em><em>, </em><em>'genbank'}</em>) – Does image use 0-based (‘biopython’) or 1-based (‘genbank’)
indexing of nucleotide sites?</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Image of target, which has <cite>.plot</cite> and <cite>.plot_with_bokeh</cite> methods:
<a class="reference external" href="https://edinburgh-genome-foundry.github.io/DnaFeaturesViewer">https://edinburgh-genome-foundry.github.io/DnaFeaturesViewer</a></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dna_features_viewer.GraphicRecord.GraphicRecord</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="alignparse.targets.Targets">
<em class="property">class </em><code class="sig-prename descclassname">alignparse.targets.</code><code class="sig-name descname">Targets</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">seqsfile</span></em>, <em class="sig-param"><span class="n">feature_parse_specs</span></em>, <em class="sig-param"><span class="n">allow_extra_features</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">seqsfileformat</span><span class="o">=</span><span class="default_value">'genbank'</span></em>, <em class="sig-param"><span class="n">allow_clipped_muts_seqs</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_feature_parse_specs_keys</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">select_target_names</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Collection of <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqsfile</strong> (<em>str</em><em> or </em><em>list</em>) – Name of file specifying the targets, or list of such files. So
if multiple targets they can all be in one file or in separate files.</p></li>
<li><p><strong>feature_parse_specs</strong> (<em>dict</em><em> or </em><em>str</em>) – <p>How <a class="reference internal" href="#alignparse.targets.Targets.parse_alignment" title="alignparse.targets.Targets.parse_alignment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Targets.parse_alignment()</span></code></a> parses alignments. Specify dict
or name of YAML file. Keyed by names of targets, values target-level
dicts keyed by feature names. The feature-level dicts have two keys:</p>
<blockquote>
<div><ul>
<li><p>’filter’: dict keyed by ‘clip5’, ‘clip3’, ‘mutation_nt_count’,
and ‘mutation_op_count’ giving max clipping at each end, number
of nucleotide mutations, and number of <code class="docutils literal notranslate"><span class="pre">cs</span></code> tag mutation
operations allowed for feature. If ‘filter’ itself or any of
the keys are missing, the value is set to zero. If the value
is <cite>None</cite> (‘null’ in YAML notation), then no filter is applied.</p></li>
<li><p>’return’: str or list of strings indicating what to return for this
feature. If ‘returns’ is absent or the value is <cite>None</cite> (‘null’ in
YAML notation), nothing is returned for this feature. Otherwise
list one or more of ‘sequence’, ‘mutations’, ‘accuracy’, ‘cs’,
‘clip5’, and ‘clip3’ to get the sequence, mutation string, <code class="docutils literal notranslate"><span class="pre">cs</span></code>
tag, or number of clipped nucleotides from each end.</p></li>
</ul>
</div></blockquote>
<p>In addition, target-level dicts should have keys ‘query_clip5’ and
‘query_clip3’ which give the max amount that can be clipped from
each end of the query prior to the alignment. Use a value of
<cite>None</cite> (‘null’ in YAML notation) to have no filter on this clipping.
Filters will be applied in the order the features appear in the
<cite>feature_parse_specs</cite>.</p>
</p></li>
<li><p><strong>allow_extra_features</strong> (<em>bool</em>) – Can targets have features not in <cite>feature_parse_specs</cite>?</p></li>
<li><p><strong>seqsfileformat</strong> (<em>{'genbank'}</em>) – Format of <cite>seqsfile</cite>. Currently, ‘genbank’ is the only supported
option. The GenBank Flat File format is described <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/samplerecord/">here</a>, but not all
fields are required. The documentation includes <a class="reference external" href="https://jbloomlab.github.io/alignparse/examples.html">examples</a> that show
what fields should typically be included. GenBank files can be readily
generated using several sequence editing programs, such as <a class="reference external" href="https://jorgensen.biology.utah.edu/wayned/ape/">ApE</a> or <a class="reference external" href="https://www.benchling.com/">Benchling</a>.</p></li>
<li><p><strong>allow_clipped_muts_seqs</strong> (<em>bool</em>) – Returning sequence or mutations for features where non-zero
clipping is allowed is dangerous, since as described in
<a class="reference internal" href="#alignparse.targets.Targets.parse_alignment" title="alignparse.targets.Targets.parse_alignment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Targets.parse_alignment()</span></code></a> these will only be for unclipped
region and so are easy to mis-interpret. So you must explicitly
set this option to <cite>True</cite> in order to allow return of mutations /
sequences for features with clipping allowed; otherwise you’ll get
an error if you try to recover such sequences / mutations.</p></li>
<li><p><strong>ignore_feature_parse_specs_keys</strong> (<em>None</em><em> or </em><em>list</em>) – Ignore these target-level keys in <cite>feature_parse_specs</cite>. Useful for
YAML with default keys that don’t represent actual targets.</p></li>
<li><p><strong>select_target_names</strong> (<em>None</em><em> or </em><em>list</em>) – If <cite>None</cite>, the created object is for all sequences in <cite>seqsfile</cite>.
Otherwise pass a list with names of just the sequences of interest.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt id="alignparse.targets.Targets.targets">
<code class="sig-name descname">targets</code><a class="headerlink" href="#alignparse.targets.Targets.targets" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Targets.target_names">
<code class="sig-name descname">target_names</code><a class="headerlink" href="#alignparse.targets.Targets.target_names" title="Permalink to this definition">¶</a></dt>
<dd><p>List of names of all targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="alignparse.targets.Targets.target_seqs">
<code class="sig-name descname">target_seqs</code><a class="headerlink" href="#alignparse.targets.Targets.target_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Keyed by target name, value is sequence as str.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.align">
<code class="sig-name descname">align</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">queryfile</span></em>, <em class="sig-param"><span class="n">alignmentfile</span></em>, <em class="sig-param"><span class="n">mapper</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.align"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.align" title="Permalink to this definition">¶</a></dt>
<dd><p>Align query sequences to targets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>queryfile</strong> (<em>str</em>) – The query sequences to align (FASTQ or FASTA, can be gzipped).</p></li>
<li><p><strong>alignmentfile</strong> (<em>str</em>) – SAM file created by <cite>mapper</cite> with alignments of queries to the
target sequences within this <a class="reference internal" href="#alignparse.targets.Targets" title="alignparse.targets.Targets"><code class="xref py py-class docutils literal notranslate"><span class="pre">Targets</span></code></a> object.</p></li>
<li><p><strong>mapper</strong> (<a class="reference internal" href="alignparse.minimap2.html#alignparse.minimap2.Mapper" title="alignparse.minimap2.Mapper"><code class="xref py py-class docutils literal notranslate"><span class="pre">alignparse.minimap2.Mapper</span></code></a>) – Mapper that runs <code class="docutils literal notranslate"><span class="pre">minimap2</span></code>. Alignment options set when creating
this mapper.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.align_and_parse">
<code class="sig-name descname">align_and_parse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df</span></em>, <em class="sig-param"><span class="n">mapper</span></em>, <em class="sig-param"><span class="n">outdir</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">name_col</span><span class="o">=</span><span class="default_value">'name'</span></em>, <em class="sig-param"><span class="n">queryfile_col</span><span class="o">=</span><span class="default_value">'queryfile'</span></em>, <em class="sig-param"><span class="n">group_cols</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">to_csv</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">overwrite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">multi_align</span><span class="o">=</span><span class="default_value">'primary'</span></em>, <em class="sig-param"><span class="n">filtered_cs</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">skip_sups</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">ncpus</span><span class="o">=</span><span class="default_value">- 1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.align_and_parse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.align_and_parse" title="Permalink to this definition">¶</a></dt>
<dd><p>Align query sequences and then parse alignments.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This is a convenience method to run <a class="reference internal" href="#alignparse.targets.Targets.align" title="alignparse.targets.Targets.align"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Targets.align()</span></code></a> and
<a class="reference internal" href="#alignparse.targets.Targets.parse_alignment" title="alignparse.targets.Targets.parse_alignment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Targets.parse_alignment()</span></code></a> on multiple queries and collate
the results.</p>
<p>It also allows multiple queries to be handled simultaneously using
multiprocessing.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame with information on queries to align.</p></li>
<li><p><strong>mapper</strong> (<a class="reference internal" href="alignparse.minimap2.html#alignparse.minimap2.Mapper" title="alignparse.minimap2.Mapper"><code class="xref py py-class docutils literal notranslate"><span class="pre">alignparse.minimap2.Mapper</span></code></a>) – Mapper that runs <code class="docutils literal notranslate"><span class="pre">minimap2</span></code>. Alignment options set when creating
this mapper.</p></li>
<li><p><strong>outdir</strong> (<em>str</em>) – Name of directory with created alignments and parsing files.
Created if it does not exist.</p></li>
<li><p><strong>name_col</strong> (<em>str</em>) – Column in <cite>df</cite> with the name of each set of queries.</p></li>
<li><p><strong>queryfile_col</strong> (<em>str</em>) – Column in <cite>df</cite> with FASTQ file with queries.</p></li>
<li><p><strong>group_cols</strong> (<cite>None</cite> or str or list) – Columns in <cite>df</cite> used to “group” results. These columns are
in all created data frames. For instance, might specify
different libraries or samples.</p></li>
<li><p><strong>to_csv</strong> (<em>bool</em>) – Write CSV files rather than return data frames. Useful to
avoid reading large data frames into memory.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – If some of the created output files already exist, do we
overwrite them or raise and error?</p></li>
<li><p><strong>multi_align</strong> (<em>{'primary'}</em>) – How to handle multiple alignments. Currently only option is
‘primary’, which ignores all secondary alignments.</p></li>
<li><p><strong>filtered_cs</strong> (<em>bool</em>) – Add <cite>cs</cite> tag that failed the filter to filtered dataframe along
with filter reason. Allows for more easily investigating why
reads are failing the filters.</p></li>
<li><p><strong>skip_sups</strong> (<em>bool</em>) – Whether or not to skip supplementary alignments when parsing.
Supplementary alignments are additional possible alignments for
a read due to the read potentially being a chimeric. The default
is to skip these alignments and <em>not</em> parse them.</p></li>
<li><p><strong>ncpus</strong> (<em>int</em>) – Number of CPUs to use; -1 means all available.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>(readstats, aligned, filtered)</strong> – Same meaning as for <a class="reference internal" href="#alignparse.targets.Targets.parse_alignment" title="alignparse.targets.Targets.parse_alignment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Targets.parse_alignment()</span></code></a> except
the data frames / CSV files all have additional columns indicating
name of each query set (<cite>name_cols</cite>) as well as any <cite>group_cols</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.feature_parse_specs">
<code class="sig-name descname">feature_parse_specs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">returntype</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.feature_parse_specs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.feature_parse_specs" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the feature parsing specs.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Filters will be applied in the order they are listed in the
<cite>feature_parse_specs</cite> <cite>yaml</cite> file or <cite>dict</cite>. Once a read fails a
filter, other filters will not be applied. As such, it is recommended
to have features with filters for 5’ and 3’ clipping listed first.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>returntype</strong> (<em>{'dict'</em><em>, </em><em>'yaml'}</em>) – Return a Python <cite>dict</cite> or a YAML string representation.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The feature parsing specs set by the <cite>feature_parse_specs</cite> at
<a class="reference internal" href="#alignparse.targets.Targets" title="alignparse.targets.Targets"><code class="xref py py-class docutils literal notranslate"><span class="pre">Targets</span></code></a> initialization, but with any missing default
values explicitly filled in.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict or str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.features_to_parse">
<code class="sig-name descname">features_to_parse</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">targetname</span></em>, <em class="sig-param"><span class="n">feature_or_name</span><span class="o">=</span><span class="default_value">'feature'</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.features_to_parse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.features_to_parse" title="Permalink to this definition">¶</a></dt>
<dd><p>Features to parse for a target.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>targetname</strong> (<em>str</em>) – Name of target.</p></li>
<li><p><strong>feature_or_name</strong> (<em>{'feature'</em><em>, </em><em>'name'}</em>) – Get the <a class="reference internal" href="#alignparse.targets.Feature" title="alignparse.targets.Feature"><code class="xref py py-class docutils literal notranslate"><span class="pre">Feature</span></code></a> objects themselves or their names.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Features to parse for this target, as specified in
<a class="reference internal" href="#alignparse.targets.Targets.feature_parse_specs" title="alignparse.targets.Targets.feature_parse_specs"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Targets.feature_parse_specs()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.get_target">
<code class="sig-name descname">get_target</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">name</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.get_target"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.get_target" title="Permalink to this definition">¶</a></dt>
<dd><p>Get <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a> by name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – Name of <a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns the target, or raises <cite>ValueError</cite> if no such target.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#alignparse.targets.Target" title="alignparse.targets.Target"><code class="xref py py-class docutils literal notranslate"><span class="pre">Target</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.parse_alignment">
<code class="sig-name descname">parse_alignment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samfile</span></em>, <em class="sig-param"><span class="n">multi_align</span><span class="o">=</span><span class="default_value">'primary'</span></em>, <em class="sig-param"><span class="n">to_csv</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">csv_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">overwrite_csv</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">filtered_cs</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">skip_sups</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.parse_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.parse_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse alignment features as specified in <cite>feature_parse_specs</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samfile</strong> (<em>str</em>) – SAM file with <code class="docutils literal notranslate"><span class="pre">minimap2</span></code> alignments with <code class="docutils literal notranslate"><span class="pre">cs</span></code> tag, typically
created by <a class="reference internal" href="#alignparse.targets.Targets.align" title="alignparse.targets.Targets.align"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Targets.align()</span></code></a>.</p></li>
<li><p><strong>multi_align</strong> (<em>{'primary'}</em>) – How to handle multiple alignments. Currently only option is
‘primary’, which ignores all secondary alignments.</p></li>
<li><p><strong>to_csv</strong> (<em>bool</em>) – Return CSV file names rather than return data frames. Useful to
avoid reading large data frames into memory.</p></li>
<li><p><strong>csv_dir</strong> (<em>None</em><em> or </em><em>str</em>) – If <cite>to_csv</cite> is <cite>True</cite>, name of directory to which we
write CSV files (created if needed). If <cite>None</cite>, write
to current directory.</p></li>
<li><p><strong>overwrite_csv</strong> (<em>bool</em>) – If using <cite>to_csv</cite>, do we overwrite existing CSV files or
raise an error if they already exist?</p></li>
<li><p><strong>filtered_cs</strong> (<em>bool</em>) – Add <cite>cs</cite> tag that failed the filter to filtered dataframe along
with filter reason. Allows for more easily investigating why
reads are failing the filters.</p></li>
<li><p><strong>skip_sups</strong> (<em>bool</em>) – Whether or not to skip supplementary alignments when parsing.
Supplementary alignments are additional possible alignments for
a read due to the read potentially being a chimeric. The default
is to skip these alignments and <em>not</em> parse them.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p><strong>(readstats, aligned, filtered)</strong> –</p>
<ul class="simple">
<li><p><cite>readstats</cite> is <cite>pandas.DataFrame</cite> with numbers of unmapped reads,
and for each target the number of mapped reads that are validly
aligned and that fail filters in <cite>feature_parse_specs</cite>.</p></li>
<li><p><cite>aligned</cite> is a dict keyed by name of each target. Entries are
<cite>pandas.DataFrame</cite> with rows for each validly aligned read. Rows
give query name, query clipping at each end of alignment, and any
feature-level info specified for return in <cite>feature_parse_specs</cite>
in columns with names equal to feature suffixed by ‘_sequence’,
‘_mutations’, ‘_accuracy’, ‘_cs’, ‘_clip5’, and ‘_clip3’.
and ‘_clip3’.</p></li>
<li><p><cite>filtered</cite> is a dict keyed by name of each target. Entries are
<cite>pandas.DataFrame</cite> with a row for each filtered aligned read
giving the query name and the reason it was filtered.
If <cite>filtered_cs</cite> is <cite>True</cite> then, add a column to the “filtered”
<cite>pandas.DataFrame`s with the `cs</cite> tag that failed the filter.</p></li>
</ul>
<p>If <cite>to_csv</cite> is <cite>True</cite>, then <cite>aligned</cite> and <cite>filtered</cite> give
names of CSV files holding data frames.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">cs</span></code> tags are in the short format returned by <code class="docutils literal notranslate"><span class="pre">minimap2</span></code>;
see here for details: <a class="reference external" href="https://lh3.github.io/minimap2/minimap2.html">https://lh3.github.io/minimap2/minimap2.html</a></p>
<p>When parsing features, if an insertion occurs between two features,
it is assigned to the end of the first feature.</p>
<p>Returned sequences, mutation strings, and <code class="docutils literal notranslate"><span class="pre">cs</span></code> tags are only for
for the portion of the feature that aligns, and do <strong>not</strong> indicate
clipping, which you instead get in the ‘_clip*’ columns. The
sequences are simply what the <code class="docutils literal notranslate"><span class="pre">cs</span></code> tag implies, indels / mutations
are not indicated in this column. Mutation strings are space-delimited
with these operations in <strong>1-based</strong> (1, 2, …) numbering from start
of the feature:</p>
<blockquote>
<div><ul class="simple">
<li><p>‘A2G’ : substitution at site 2 from A to G</p></li>
<li><p>‘ins5TAA’ : insertion of ‘TAA’ starting at site 5</p></li>
<li><p>‘del5to6’ : deletion of sites 5 to 6, inclusive</p></li>
</ul>
</div></blockquote>
<p>The returned accuracy is the average accuracy of the aligned
query sites as calculated from the Q-values, and is <cite>nan</cite> if
there are no aligned query sites.</p>
</div>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.plot">
<code class="sig-name descname">plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">sharex</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">ax_width</span><span class="o">=</span><span class="default_value">5</span></em>, <em class="sig-param"><span class="n">ax_height</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">hspace</span><span class="o">=</span><span class="default_value">0.4</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot all the targets.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For more customizable plots, call <a class="reference internal" href="#alignparse.targets.Target.image" title="alignparse.targets.Target.image"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Target.image()</span></code></a> for individual
targets.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sharex</strong> (<em>bool</em>) – Share x-axis among plots for each target?</p></li>
<li><p><strong>ax_width</strong> (<em>float</em>) – Width of each axis in inches.</p></li>
<li><p><strong>ax_height</strong> (<em>float</em>) – Height of each axis in inches.</p></li>
<li><p><strong>hspace</strong> (<em>float</em>) – Vertical space between axes as fraction of <cite>ax_height</cite>.</p></li>
<li><p><strong>**kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#alignparse.targets.Target.image" title="alignparse.targets.Target.image"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Target.image()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Figure showing all targets.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>matplotlib.pyplot.figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="alignparse.targets.Targets.write_fasta">
<code class="sig-name descname">write_fasta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fastafile</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/targets.html#Targets.write_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.targets.Targets.write_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Write all targets to a FASTA file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>str</em><em> or </em><em>writable file-like object</em>) – Write targets to this file.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">alignparse</h1>
    
  </a>
</p>



<p class="blurb">Align sequences and then parse features.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=alignparse&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/alignparse">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/alignparse.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/alignparse.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">alignparse documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="alignparse.html">alignparse package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="alignparse.html#alignparse">alignparse</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="alignparse.html#submodules">Submodules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">Package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="alignparse.html">alignparse package</a><ul>
      <li>Previous: <a href="alignparse.minimap2.html" title="previous chapter">minimap2</a></li>
      <li>Next: <a href="alignparse.utils.html" title="next chapter">utils</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2020.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/alignparse.targets.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/alignparse" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>