
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>consensus &#8212; alignparse 0.1.3 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="constants" href="alignparse.constants.html" />
    <link rel="prev" title="ccs" href="alignparse.ccs.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-alignparse.consensus"></span><div class="section" id="consensus">
<h1>consensus<a class="headerlink" href="#consensus" title="Permalink to this headline">¶</a></h1>
<p>Get consensus of sequences / mutations. Useful for grouping sequences
within a barcode.</p>
<dl class="class">
<dt id="alignparse.consensus.Mutations">
<em class="property">class </em><code class="sig-prename descclassname">alignparse.consensus.</code><code class="sig-name descname">Mutations</code><span class="sig-paren">(</span><em class="sig-param">substitutions</em>, <em class="sig-param">deletions</em>, <em class="sig-param">insertions</em><span class="sig-paren">)</span><a class="headerlink" href="#alignparse.consensus.Mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="method">
<dt id="alignparse.consensus.Mutations.deletions">
<em class="property">property </em><code class="sig-name descname">deletions</code><a class="headerlink" href="#alignparse.consensus.Mutations.deletions" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="method">
<dt id="alignparse.consensus.Mutations.insertions">
<em class="property">property </em><code class="sig-name descname">insertions</code><a class="headerlink" href="#alignparse.consensus.Mutations.insertions" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="method">
<dt id="alignparse.consensus.Mutations.substitutions">
<em class="property">property </em><code class="sig-name descname">substitutions</code><a class="headerlink" href="#alignparse.consensus.Mutations.substitutions" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="alignparse.consensus.add_mut_info_cols">
<code class="sig-prename descclassname">alignparse.consensus.</code><code class="sig-name descname">add_mut_info_cols</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">*</em>, <em class="sig-param">mutation_col='mutations'</em>, <em class="sig-param">sub_str_col=None</em>, <em class="sig-param">del_str_col=None</em>, <em class="sig-param">ins_str_col=None</em>, <em class="sig-param">indel_str_col=None</em>, <em class="sig-param">n_sub_col=None</em>, <em class="sig-param">n_del_col=None</em>, <em class="sig-param">n_ins_col=None</em>, <em class="sig-param">n_indel_col=None</em>, <em class="sig-param">overwrite_cols=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/consensus.html#add_mut_info_cols"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.consensus.add_mut_info_cols" title="Permalink to this definition">¶</a></dt>
<dd><p>Expand information about mutations in a data frame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame with each row giving information about a variant.</p></li>
<li><p><strong>mutation_col</strong> (<em>str</em>) – Name of column in <cite>df</cite> with str containing space-delimited
list of mutations in format parseable by <a class="reference internal" href="#alignparse.consensus.process_mut_str" title="alignparse.consensus.process_mut_str"><code class="xref py py-func docutils literal notranslate"><span class="pre">process_mut_str()</span></code></a>.</p></li>
<li><p><strong>sub_str_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with string giving substitutions.</p></li>
<li><p><strong>del_str_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with string giving deletions.</p></li>
<li><p><strong>ins_str_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with string giving insertions.</p></li>
<li><p><strong>indel_str_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with string giving deletions and insertions.</p></li>
<li><p><strong>n_sub_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with number of substitutions.</p></li>
<li><p><strong>n_del_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with number of deletions.</p></li>
<li><p><strong>n_ins_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with number of insertions.</p></li>
<li><p><strong>n_indel_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with number of deletions and insertions.</p></li>
<li><p><strong>overwrite_cols</strong> (<em>bool</em>) – If column to be created is already in <cite>df</cite>, overwrite it?</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <strong>copy</strong> of <cite>df</cite> with the indicated columns added.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Data frame for which we get mutation info:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>      <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;seq3&#39;</span><span class="p">],</span>
<span class="gp">... </span>      <span class="s1">&#39;mutations&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;A6G del7to8 T3A&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;T5A ins5AAT del8to9&#39;</span><span class="p">],</span>
<span class="gp">... </span>      <span class="p">})</span>
</pre></div>
</div>
<p>Get info on substitutions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">add_mut_info_cols</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">sub_str_col</span><span class="o">=</span><span class="s1">&#39;subs&#39;</span><span class="p">,</span> <span class="n">n_sub_col</span><span class="o">=</span><span class="s1">&#39;nsubs&#39;</span><span class="p">)</span>
<span class="go">   name            mutations     subs  nsubs</span>
<span class="go">0  seq1      A6G del7to8 T3A  T3A A6G      2</span>
<span class="go">1  seq2                                    0</span>
<span class="go">2  seq3  T5A ins5AAT del8to9      T5A      1</span>
</pre></div>
</div>
<p>Get info on substitutions and indels:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;max_columns&#39;</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">add_mut_info_cols</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">sub_str_col</span><span class="o">=</span><span class="s1">&#39;subs&#39;</span><span class="p">,</span> <span class="n">n_sub_col</span><span class="o">=</span><span class="s1">&#39;nsubs&#39;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">indel_str_col</span><span class="o">=</span><span class="s1">&#39;indels&#39;</span><span class="p">,</span> <span class="n">n_indel_col</span><span class="o">=</span><span class="s1">&#39;nindels&#39;</span><span class="p">)</span>
<span class="go">   name            mutations     subs           indels  nsubs  nindels</span>
<span class="go">0  seq1      A6G del7to8 T3A  T3A A6G          del7to8      2        1</span>
<span class="go">1  seq2                                                     0        0</span>
<span class="go">2  seq3  T5A ins5AAT del8to9      T5A  del8to9 ins5AAT      1        2</span>
</pre></div>
</div>
<p>Just get counts for all types of mutations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">add_mut_info_cols</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">n_sub_col</span><span class="o">=</span><span class="s1">&#39;nsubs&#39;</span><span class="p">,</span> <span class="n">n_del_col</span><span class="o">=</span><span class="s1">&#39;ndels&#39;</span><span class="p">,</span>
<span class="gp">... </span>                  <span class="n">n_ins_col</span><span class="o">=</span><span class="s1">&#39;nins&#39;</span><span class="p">,</span> <span class="n">n_indel_col</span><span class="o">=</span><span class="s1">&#39;nindels&#39;</span><span class="p">)</span>
<span class="go">   name            mutations  nsubs  ndels  nins  nindels</span>
<span class="go">0  seq1      A6G del7to8 T3A      2      1     0        1</span>
<span class="go">1  seq2                           0      0     0        0</span>
<span class="go">2  seq3  T5A ins5AAT del8to9      1      1     1        2</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="alignparse.consensus.empirical_accuracy">
<code class="sig-prename descclassname">alignparse.consensus.</code><code class="sig-name descname">empirical_accuracy</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">*</em>, <em class="sig-param">group_cols='barcode'</em>, <em class="sig-param">upstream_group_cols='library'</em>, <em class="sig-param">mutation_col='mutations'</em>, <em class="sig-param">accuracy_col='accuracy'</em>, <em class="sig-param">sort_mutations=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/consensus.html#empirical_accuracy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.consensus.empirical_accuracy" title="Permalink to this definition">¶</a></dt>
<dd><p>Accuracy from number of identical sequences in a group (i.e., barcode).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The accuracy <span class="math notranslate nohighlight">\(1 - \epsilon\)</span> is calculated as follows:</p>
<p>Let <span class="math notranslate nohighlight">\(\epsilon\)</span> be the probability a sequence has an error.
Consider a group of <span class="math notranslate nohighlight">\(n\)</span> sequences that should be identical in
the absence of errors (i.e., they all have the same barcode), and
assume errors only make sequences <strong>different</strong> (the chance of two
sequences having the same error is very low). We would like to
calculate the probability of observing <span class="math notranslate nohighlight">\(u\)</span> unique sequences in
the group given <span class="math notranslate nohighlight">\(\epsilon\)</span> and <span class="math notranslate nohighlight">\(n\)</span>.</p>
<p>First, consider the case where there are two sequences in the group
(<span class="math notranslate nohighlight">\(n = 2\)</span>). The probability of having <span class="math notranslate nohighlight">\(u = 1\)</span> unique
sequences in the group is the probability that neither have an error:</p>
<div class="math notranslate nohighlight">
\[\Pr\left(u=1 | n=2, \epsilon\right) = \left(1 - \epsilon\right)^2.\]</div>
<p>The probability of having <span class="math notranslate nohighlight">\(u = 2\)</span> unique sequences is the
the probability that either one or both have errors:</p>
<div class="math notranslate nohighlight">
\[\Pr\left(u=2 | n=2, \epsilon\right) = 2\left(1 - \epsilon\right)\epsilon
+ \epsilon^2.\]</div>
<p>Generalizing to arbitrary <span class="math notranslate nohighlight">\(n\)</span>, we have:</p>
<div class="math notranslate nohighlight">
\[\Pr\left(u | n, \epsilon\right) = \binom{n}{u-1}
\left(1 - \epsilon\right)^{n - u + 1} \epsilon^{u - 1} +
\delta_{un} \epsilon^n\]</div>
<p>where <span class="math notranslate nohighlight">\(\binom{n}{u-1}\)</span> is the binomial coefficient and
<span class="math notranslate nohighlight">\(\delta_{un}\)</span> is the Kronecker delta.</p>
<p>Let there be <span class="math notranslate nohighlight">\(G\)</span> groups of sequences, with the size and number
of unique sequences in group <span class="math notranslate nohighlight">\(g\)</span> be <span class="math notranslate nohighlight">\(n_g\)</span> and <span class="math notranslate nohighlight">\(u_g\)</span>
(where <span class="math notranslate nohighlight">\(g = 1, \ldots, G\)</span>).
The overall likelihood <span class="math notranslate nohighlight">\(L\)</span> of the observed numbers of unique
sequences given the group sizes and error rate is:</p>
<div class="math notranslate nohighlight">
\[\begin{split}L &amp;=&amp; \Pr\left(\left\{u_g\right\}|\left\{n_g\right\},\epsilon\right) \\
&amp;=&amp; \prod_g \Pr\left(u_g | n_g, \epsilon\right).\end{split}\]</div>
<p>To find the maximum likelihood of error rate, we simply use numerical
optimization to find the value of <span class="math notranslate nohighlight">\(\epsilon\)</span> that maximizes
<span class="math notranslate nohighlight">\(L\)</span>. In practice, we actually maximize <span class="math notranslate nohighlight">\(\ln\left(L\right)\)</span>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Each row gives data for a different sequence.</p></li>
<li><p><strong>group_cols</strong> (<em>str</em><em> or </em><em>list</em>) – Column(s) in <cite>df</cite> indicating how we group sequences before computing
fraction within group that is identical.</p></li>
<li><p><strong>upstream_group_cols</strong> (<em>str</em><em> or </em><em>None</em><em> or </em><em>list</em>) – Column(s) in <cite>df</cite> that we use to group <strong>before</strong> doing the grouping
for computing fraction identical. So a different average fraction
will be returned for each of these upstream groups.</p></li>
<li><p><strong>mutation_col</strong> (<em>str</em>) – Column in <cite>df</cite> that we compare when determining if sequences are
identical. Typically list of mutations or sequences themselves.</p></li>
<li><p><strong>accuracy_col</strong> (<em>str</em>) – Name of column in returned data frame that gives empirical accuracy.</p></li>
<li><p><strong>sort_mutations</strong> (<em>bool</em>) – Useful if you have strings of space-delimited mutations not guaranteed
to be consistently ordered. If <cite>True</cite>, sort such space-delimited
mutations.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Has all columns in <cite>upstream_group_cols</cite> plus the new column with
name given by <cite>accuracy_col</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>        <span class="s1">&#39;barcode&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">],</span>
<span class="gp">... </span>        <span class="s1">&#39;mutations&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;A1G&#39;</span><span class="p">,</span> <span class="s1">&#39;A1G&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;T2A C6G&#39;</span><span class="p">,</span> <span class="s1">&#39;T5C&#39;</span><span class="p">,</span> <span class="s1">&#39;C6G T2A&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;precision&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">empirical_accuracy</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">upstream_group_cols</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="go">   accuracy</span>
<span class="go">0    0.7692</span>
</pre></div>
</div>
<p>If we do the same without sorting the mutations, we get lower values
as ‘T2A C6G’ and ‘C6G T2A’ are then considered as different:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;precision&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">empirical_accuracy</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">upstream_group_cols</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>                       <span class="n">sort_mutations</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">   accuracy</span>
<span class="go">0    0.4766</span>
</pre></div>
</div>
<p>Now another example with two libraries and non-standard column
names. Note that we only get results for the two libraries
with barcodes found multiple times:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
<span class="gp">... </span>    <span class="s1">&#39;bc&#39;</span>  <span class="p">:[</span><span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;AT&#39;</span><span class="p">,</span> <span class="s1">&#39;TG&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">,</span> <span class="s1">&#39;TA&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;var&#39;</span> <span class="p">:[</span><span class="s1">&#39;v1&#39;</span><span class="p">,</span> <span class="s1">&#39;v1&#39;</span><span class="p">,</span> <span class="s1">&#39;v2&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">,</span> <span class="s1">&#39;v4&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">,</span> <span class="s1">&#39;v3&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;lib&#39;</span> <span class="p">:[</span><span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;s1&#39;</span><span class="p">,</span> <span class="s1">&#39;s2&#39;</span><span class="p">,</span> <span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;s3&#39;</span><span class="p">,</span> <span class="s1">&#39;s4&#39;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">pd</span><span class="o">.</span><span class="n">option_context</span><span class="p">(</span><span class="s1">&#39;precision&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">empirical_accuracy</span><span class="p">(</span><span class="n">df2</span><span class="p">,</span> <span class="n">upstream_group_cols</span><span class="o">=</span><span class="s1">&#39;lib&#39;</span><span class="p">,</span> <span class="n">group_cols</span><span class="o">=</span><span class="s1">&#39;bc&#39;</span><span class="p">,</span>
<span class="gp">... </span>                       <span class="n">mutation_col</span><span class="o">=</span><span class="s1">&#39;var&#39;</span><span class="p">,</span> <span class="n">sort_mutations</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">  lib  accuracy</span>
<span class="go">0  s1    1.0000</span>
<span class="go">1  s3    0.6667</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="alignparse.consensus.process_mut_str">
<code class="sig-prename descclassname">alignparse.consensus.</code><code class="sig-name descname">process_mut_str</code><span class="sig-paren">(</span><em class="sig-param">s</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/consensus.html#process_mut_str"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.consensus.process_mut_str" title="Permalink to this definition">¶</a></dt>
<dd><p>Process a string of mutations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>s</strong> (<em>str</em>) – Space-delimited mutations. Substitutions in form ‘A6T’. Deletions
in form ‘del5to7’. Insertions in form ‘ins6len2’ or ‘ins6TA’.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>The tuple is named <cite>Mutations</cite> and has the following elements in order:</dt><dd><ul class="simple">
<li><p><cite>substitutions</cite>: list of substitutions in order of sequence.</p></li>
<li><p><cite>deletions</cite>: list of deletions in order of sequence.</p></li>
<li><p><cite>insertions</cite>: list of insertions in order of sequence.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>namedtuple</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;A1T del5to7 G9C ins6len2 del12to12&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">process_mut_str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>  
<span class="go">Mutations(substitutions=[&#39;A1T&#39;, &#39;G9C&#39;],</span>
<span class="go">          deletions=[&#39;del5to7&#39;, &#39;del12to12&#39;],</span>
<span class="go">          insertions=[&#39;ins6len2&#39;])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="s1">&#39;A11T del5to7 G9C ins6GA del12to15 ins13AAT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">process_mut_str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>  
<span class="go">Mutations(substitutions=[&#39;G9C&#39;, &#39;A11T&#39;],</span>
<span class="go">          deletions=[&#39;del5to7&#39;, &#39;del12to15&#39;],</span>
<span class="go">          insertions=[&#39;ins6GA&#39;, &#39;ins13AAT&#39;])</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="alignparse.consensus.simple_mutconsensus">
<code class="sig-prename descclassname">alignparse.consensus.</code><code class="sig-name descname">simple_mutconsensus</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">*</em>, <em class="sig-param">group_cols=('library'</em>, <em class="sig-param">'barcode')</em>, <em class="sig-param">mutation_col='mutations'</em>, <em class="sig-param">max_sub_diffs=1</em>, <em class="sig-param">max_indel_diffs=2</em>, <em class="sig-param">max_minor_sub_frac=0.1</em>, <em class="sig-param">max_minor_indel_frac=0.25</em>, <em class="sig-param">support_col='variant_call_support'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/alignparse/consensus.html#simple_mutconsensus"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#alignparse.consensus.simple_mutconsensus" title="Permalink to this definition">¶</a></dt>
<dd><p>Get simple consensus of mutations with group (i.e., barcode).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Each row gives data on mutations for a different sequence.</p></li>
<li><p><strong>group_cols</strong> (<em>list</em><em>, </em><em>tuple</em><em>, or </em><em>str</em>) – Group all sequences that share values in these column(s) of <cite>df</cite>.</p></li>
<li><p><strong>mutation_col</strong> (<em>str</em>) – Column in <cite>df</cite> with mutations in form that can be processed by
<code class="xref py py-func docutils literal notranslate"><span class="pre">processe_mut_str()</span></code>.</p></li>
<li><p><strong>max_sub_diffs</strong> (<em>int</em>) – Drop groups where any variant differs from all other variants
by more than this many substitution (point mutation) differences.</p></li>
<li><p><strong>max_indel_diffs</strong> (<em>int</em>) – Drop groups where any variant differs from all other variants
by more than this many indel differences.</p></li>
<li><p><strong>max_minor_sub_frac</strong> (<em>float</em>) – Drop groups with a minor (non-consensus) substitution in more than the
<strong>ceiling</strong> of this fraction times the number of sequences in group.</p></li>
<li><p><strong>max_minor_indel_frac</strong> (<em>float</em>) – Drop groups with a minor (non-consensus) indel in more than the
<strong>ceiling</strong> of this fraction times the number of sequences in group.</p></li>
<li><p><strong>support_col</strong> (<em>str</em>) – Name of column in returned <cite>consensus</cite> data frame with number
of sequences supporting the consensus call.</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The rationale behind this consensus calling scheme is that we want
to build a consensus except in the following two cases, each of which
indicate a likely problem <strong>beyond</strong> simple rare sequencing errors (such
as strand exchange or barcode collisions):</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>A sequence within the group differs by too much from others in
group (given by <cite>max_sub_diffs</cite> and <cite>max_indel_diffs</cite>).</p></li>
<li><p>There are “minor” (non-consensus) mutations present at too high
frequency (given by <cite>max_minor_indel_frac</cite> and <cite>max_minor_sub_frac</cite>).</p></li>
</ol>
</div></blockquote>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>The returned tuple <cite>(consensus, dropped)</cite> consists of two
pandas data frames:</p>
<blockquote>
<div><ul class="simple">
<li><p><cite>consensus</cite>: Gives consensus for all groups for which
this can be constructed. Columns are <cite>group_cols</cite>, column
with name given by <cite>mutation_col</cite> that gives consensus mutations,
and a column with name given by <cite>support_col</cite> that gives the
number of variants used to generate the consensus.</p></li>
<li><p><cite>dropped</cite>: Gives each dropped group, the reason it was dropped,
and the number of variants (sequences) in that group.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Create a dummy CSV file with our variants and mutations:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">textwrap</span><span class="o">.</span><span class="n">dedent</span><span class="p">(</span><span class="s1">&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="s1">       library,barcode,mutations</span>
<span class="gp">... </span><span class="s1">       lib_1,AG,A2C del5to7</span>
<span class="gp">... </span><span class="s1">       lib_1,AG,A2C</span>
<span class="gp">... </span><span class="s1">       lib_1,TA,G3A ins4len3</span>
<span class="gp">... </span><span class="s1">       lib_2,TA,C5A T6C</span>
<span class="gp">... </span><span class="s1">       lib_2,TA,ins5len1 T6C</span>
<span class="gp">... </span><span class="s1">       lib_2,TA,T6C</span>
<span class="gp">... </span><span class="s1">       lib_2,TA,A4G T6C</span>
<span class="gp">... </span><span class="s1">       lib_2,TG,T6A</span>
<span class="gp">... </span><span class="s1">       lib_2,TG,A2G</span>
<span class="gp">... </span><span class="s1">       lib_2,GG,del1to4</span>
<span class="gp">... </span><span class="s1">       lib_2,GG,A1C</span>
<span class="gp">... </span><span class="s1">       lib_2,AA,</span>
<span class="gp">... </span><span class="s1">       lib_2,AA,</span>
<span class="gp">... </span><span class="s1">       lib_2,AA,T6C</span>
<span class="gp">... </span><span class="s1">       lib_2,AA,T6C</span>
<span class="gp">... </span><span class="s1">       lib_3,AA,ins1len1 del1to2 T6G</span>
<span class="gp">... </span><span class="s1">       lib_3,AA,del5to7 T6G</span>
<span class="gp">... </span><span class="s1">       &#39;&#39;&#39;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Read into data frame. Note use of <cite>na_filter=False</cite> so empty strings
are read as such rather than as <cite>na</cite> values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">na_filter</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>Get the consensus mutations, and the dropped libraries / barcodes:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">consensus</span><span class="p">,</span> <span class="n">dropped</span> <span class="o">=</span> <span class="n">simple_mutconsensus</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">consensus</span>
<span class="go">  library barcode     mutations  variant_call_support</span>
<span class="go">0   lib_1      AG           A2C                     2</span>
<span class="go">1   lib_1      TA  G3A ins4len3                     1</span>
<span class="go">2   lib_2      GG                                   2</span>
<span class="go">3   lib_2      TA           T6C                     4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dropped</span>
<span class="go">  library barcode              drop_reason  nseqs</span>
<span class="go">0   lib_2      AA  minor subs too frequent      4</span>
<span class="go">1   lib_2      TG      subs diff too large      2</span>
<span class="go">2   lib_3      AA    indels diff too large      2</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">alignparse</h1>
    
  </a>
</p>



<p class="blurb">Align sequences and then parse features.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=alignparse&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/alignparse">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/alignparse.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/alignparse.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">alignparse documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="alignparse.html">alignparse package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="alignparse.html#alignparse">alignparse</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="alignparse.html#submodules">Submodules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">Package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="alignparse.html">alignparse package</a><ul>
      <li>Previous: <a href="alignparse.ccs.html" title="previous chapter">ccs</a></li>
      <li>Next: <a href="alignparse.constants.html" title="next chapter">constants</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2020.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/alignparse.consensus.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/alignparse" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>