<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Single-cell virus sequencing &#8212; alignparse 0.7.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=877b1dcb" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css?v=7f9a90b1" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />
    <script src="_static/documentation_options.js?v=fe7df9b0"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="alignparse package" href="alignparse.html" />
    <link rel="prev" title="Lassa virus glycoprotein PacBio sequencing" href="lasv_pilot.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="Single-cell-virus-sequencing">
<h1>Single-cell virus sequencing<a class="headerlink" href="#Single-cell-virus-sequencing" title="Link to this heading">¶</a></h1>
<p>This example shows how to use <a class="reference external" href="https://jbloomlab.github.io/alignparse/">alignparse</a> to process full-length PacBio CCSs of viral cDNAs generated in single-cell virus sequencing of influenza-infected cells. Specifically, it processes a snippet of data taking from <a class="reference external" href="https://jvi.asm.org/content/93/14/e00500-19">Russell et al, 2019</a>, which was generated by PacBio sequencing of viral cDNAs with unique molecular identifiers (UMIS) and cell barcodes added using a 10X Chromium (v2 reagents). This
notebook aligns these CCSs to the influenza transcripts and parses the barcodes / UMIs and viral mutations.</p>
<section id="Set-up-for-analysis">
<h2>Set up for analysis<a class="headerlink" href="#Set-up-for-analysis" title="Link to this heading">¶</a></h2>
<p>Import necessary Python modules. We use <a class="reference external" href="https://jbloomlab.github.io/alignparse/">alignparse</a> for most of the operations, <a class="reference external" href="https://plotnine.readthedocs.io">plotnine</a> for ggplot2-like plotting:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">plotnine</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.ccs</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.consensus</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.minimap2</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.targets</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">alignparse.constants</span><span class="w"> </span><span class="kn">import</span> <span class="n">CBPALETTE</span>
</pre></div>
</div>
</div>
<p>Suppress warnings that clutter output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Directory for output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outdir</span> <span class="o">=</span> <span class="s2">&quot;./output_files/&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Target-amplicons">
<h2>Target amplicons<a class="headerlink" href="#Target-amplicons" title="Link to this heading">¶</a></h2>
<p>The alignment targets consist of PCR amplicons covering the influenza virus mRNAs (WSN strain) with a 5’ termini corresponding to the primer binding site, the 3’ polyA tail used for reverse transcription, the UMI and cell barcode, and the common 3’ termini. There are also variant tags distinguishing the synonymously tagged viral variants (see <a class="reference external" href="https://jvi.asm.org/content/93/14/e00500-19">Russell et al, 2019</a>).</p>
<p>First, let’s look at the Genbank file holding the targets. The targets are defined in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/samplerecord/">Genbank Flat File format</a> Below we show the first 85 lines (the actual file is quite large as it defined all 10 influenza mRNAs):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">targetfile</span> <span class="o">=</span> <span class="s2">&quot;input_files/flu_WSN_amplicons.gb&quot;</span>

<span class="n">nlines_to_show</span> <span class="o">=</span> <span class="mi">85</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">targetfile</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">nlines_to_show</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LOCUS       fluPB2                  2395 bp    DNA              UNK 01-JAN-1980
DEFINITION  fluPB2.
ACCESSION   fluPB2
VERSION     fluPB2
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     termini5        1..34
     variant_tag_1   143..143
     variant_tag_2   149..149
     variant_tag_3   2163..2163
     variant_tag_4   2171..2171
     sequenced_mRNA  35..2316
     final_mRNA_nts  2317..2319
     polyA           2320..2347
     UMI             2348..2357
     cellbarcode     2358..2373
     termini3        2374..2395
ORIGIN
        1 gcgaaagcag gtcaattata ttcaatatgg aaagaataaa agaactaagg aatctaatgt
       61 cgcagtctcg cactcgcgag atactcacaa aaaccaccgt ggaccatatg gccataatca
      121 agaagtacac atcaggaaga cargagaara acccagcact taggatgaaa tggatgatgg
      181 caatgaaata tccaattaca gcagacaaga ggataacgga aatgattcct gagagaaatg
      241 agcagggaca aactttatgg agtaaaatga atgacgccgg atcagaccga gtgatggtat
      301 cacctctggc tgtgacatgg tggaatagga atggaccagt gacaagtaca gttcattatc
      361 caaaaatcta caaaacttat tttgaaaaag tcgaaaggtt aaaacatgga acctttggcc
      421 ctgtccattt tagaaaccaa gtcaaaatac gtcgaagagt tgacataaat cctggtcatg
      481 cagatctcag tgccaaagag gcacaggatg taatcatgga agttgttttc cctaacgaag
      541 tgggagccag gatactaaca tcggaatcgc aactaacgac aaccaaagag aagaaagaag
      601 aactccaggg ttgcaaaatt tctcctctga tggtggcata catgttggag agagaactgg
      661 tccgcaaaac gagattcctc ccagtggctg gtggaacaag cagtgtgtac attgaagtgt
      721 tgcatttgac ccaaggaaca tgctgggaac agatgtacac tccaggaggg gaggcgagga
      781 atgatgatgt tgatcaaagc ttaattattg ctgctagaaa catagtaaga agagccacag
      841 tatcagcaga tccactagca tctttattgg agatgtgcca cagcacgcag attggtggaa
      901 taaggatggt aaacatcctt aggcagaacc caacagaaga gcaagccgtg gatatttgca
      961 aggctgcaat gggactgaga attagctcat ccttcagttt tggtggattc acatttaaga
     1021 gaacaagcgg atcatcagtc aagagagagg aagaggtgct tacgggcaat cttcagacat
     1081 tgaagataag agtgcatgag ggatatgaag agttcacaat ggttgggaga agagcaacag
     1141 ctatactcag aaaagcaacc aggagattga ttcagctgat agtgagtggg agagacgaac
     1201 agtcgattgc cgaagcaata attgtggcca tggtattttc acaagaggat tgtatgataa
     1261 aagcagttag aggtgacctg aatttcgtca atagggcgaa tcagcgattg aatcccatgc
     1321 accaactttt gagacatttt cagaaggatg caaaggtgct ctttcaaaat tggggaattg
     1381 aatccatcga caatgtgatg ggaatgatcg ggatattgcc cgacatgact ccaagcaccg
     1441 agatgtcaat gagaggagtg agaatcagca aaatgggggt agatgagtat tccagcgcgg
     1501 agaagatagt ggtgagcatt gaccgttttt tgagagttag ggaccaacgt gggaatgtac
     1561 tactgtctcc cgaggagatc agtgaaacac agggaacaga gaaactgaca ataacttact
     1621 catcgtcaat gatgtgggag attaatggtc ctgaatcagt gttggtcaat acctatcagt
     1681 ggatcatcag aaactgggaa actgttaaaa ttcagtggtc ccagaatcct acaatgctgt
     1741 acaataaaat ggaatttgag ccatttcagt ctttagttcc aaaggccgtt agaggccaat
     1801 acagtgggtt tgtgagaact ctgttccaac aaatgaggga tgtgcttggg acatttgata
     1861 ccgctcagat aataaaactt cttcccttcg cagccgctcc accaaagcaa agtagaacgc
     1921 agttctcctc attgactata aatgtgaggg gatcaggaat gagaatactt gtaaggggca
     1981 attctccagt attcaactac aacaagacca ctaaaagact cacagttctc ggaaaggatg
     2041 ctggcccttt aactgaagac ccagatgaag gcacagctgg agttgagtcc gcagttctga
     2101 gaggattcct cattctgggc aaagaagaca ggagatatgg accagcatta agcataaatg
     2161 aaytgagcaa ycttgcgaaa ggagagaagg ctaatgtgct aattgggcaa ggagacgtgg
     2221 tgttggtaat gaaacggaaa cggaactcta gcatacttac tgacagccag acagcgacca
     2281 aaagaattcg gatggccatc aattagtgtc gaatagttta aaaaaaaaaa aaaaaaaaaa
     2341 aaaaaaannn nnnnnnnnnn nnnnnnnnnn nnnagatcgg aagagcgtcg tgtag
//
LOCUS       fluPB1                  2395 bp    DNA              UNK 01-JAN-1980
DEFINITION  fluPB1.
ACCESSION   fluPB1
VERSION     fluPB1
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     termini5        1..22
     variant_tag_1   146..146
     variant_tag_2   152..152
     variant_tag_3   2159..2159
     variant_tag_4   2162..2162
     sequenced_mRNA  23..2316
     final_mRNA_nts  2317..2319
     polyA           2320..2347
     UMI             2348..2357
     cellbarcode     2358..2373
     termini3        2374..2395
ORIGIN
        1 gcgaaagcag gcaaaccatt tgaatggatg tcaatccgac tttacttttc ttaaaagtgc
       61 cagcacaaaa tgctataagc acaactttcc cttatactgg agaccctcct tacagccatg

</pre></div></div>
</div>
<p>We also have the YAML file specifying how to parse the features in the alignments. Note how this file uses the YAML syntax for defaults to avoid repeating the same specifications for each target:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">feature_parse_specs_file</span> <span class="o">=</span> <span class="s2">&quot;input_files/flu_WSN_feature_parse_specs.yaml&quot;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">feature_parse_specs_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
# default for genes with two variant tags
default_2tags: &amp;default_2tags
  query_clip5: 4
  query_clip3: 4
  termini5:
    filter:
      clip5: 4
      mutation_nt_count: 1
      mutation_op_count: null
  sequenced_mRNA:
    filter:
      mutation_nt_count: null
      mutation_op_count: 10
    return: [mutations, accuracy]
  final_mRNA_nts:
    filter:
      mutation_nt_count: null
      mutation_op_count: 2
  polyA:
    filter:
      mutation_nt_count: 12
      mutation_op_count: 4
  UMI:
    return: [sequence, accuracy]
  cellbarcode:
    return: [sequence, accuracy]
  termini3:
    filter:
      clip5: 4
      mutation_nt_count: 1
      mutation_op_count: null
  variant_tag_1:
    filter:
      mutation_nt_count: 1
      mutation_op_count: null
    return: sequence
  variant_tag_2:
    filter:
      mutation_nt_count: 1
      mutation_op_count: null
    return: sequence

# default for genes with four variant tags
default_4tags: &amp;default_4tags
  &lt;&lt;: *default_2tags
  variant_tag_3:
    filter:
      mutation_nt_count: 1
      mutation_op_count: null
    return: sequence
  variant_tag_4:
    filter:
      mutation_nt_count: 1
      mutation_op_count: null
    return: sequence

# use defaults to define specs for actual genes
fluPB2:
  &lt;&lt;: *default_4tags

fluPB1:
  &lt;&lt;: *default_4tags

fluPA:
  &lt;&lt;: *default_4tags

fluHA:
  &lt;&lt;: *default_4tags

fluNP:
  &lt;&lt;: *default_4tags

fluNA:
  &lt;&lt;: *default_4tags

fluM1:
  &lt;&lt;: *default_4tags

fluM2:
  &lt;&lt;: *default_2tags

fluNS1:
  &lt;&lt;: *default_4tags

fluNS2:
  &lt;&lt;: *default_2tags

</pre></div></div>
</div>
<p>We now read the targets into an <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">alignparse.targets.Targets</a> object with the feature-parsing specs (ignoring the keys that define thet defaults):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">targets</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">targets</span><span class="o">.</span><span class="n">Targets</span><span class="p">(</span>
    <span class="n">seqsfile</span><span class="o">=</span><span class="n">targetfile</span><span class="p">,</span>
    <span class="n">feature_parse_specs</span><span class="o">=</span><span class="n">feature_parse_specs_file</span><span class="p">,</span>
    <span class="n">ignore_feature_parse_specs_keys</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;default_2tags&quot;</span><span class="p">,</span> <span class="s2">&quot;default_4tags&quot;</span><span class="p">],</span>
    <span class="n">allow_extra_features</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">Targets</a>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax_width</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/flu_virus_seq_example_14_0.png" src="_images/flu_virus_seq_example_14_0.png" />
</div>
</div>
<p>Several things of note about how these targets are defined. These all relate to the fact that the most difficult region to parse is the polyA tail and the UMI / cell barcode–the reason being that the polyA tail is not sequenced very accurately (it is a homopolymer) plus appears to be somewhat impure in the actual primers, so is difficult to align correctly. This is compounded by the fact that we don’t know the actual UMI sequence and so don’t want indels of the polyA tail to “bleed” into the
UMI or gene alignment. Therefore:</p>
<ul class="simple">
<li><p>We define a separate feature for the final nucleotides (last 3) of the mRNA upstream of the sequenced mRNA, so polyA indels don’t get assigned to the sequenced mRNA too often.</p></li>
<li><p>We define the polyA to be just 28 nucleotides in the alignment even though the length in the primer is 30, so that deletions don’t affect the UMI.</p></li>
</ul>
<p>These empirically help with the alilgnments.</p>
</section>
<section id="PacBio-CCSs">
<h2>PacBio CCSs<a class="headerlink" href="#PacBio-CCSs" title="Link to this heading">¶</a></h2>
<p>Now let’s look at the PacBio CCSs. First, define a data frame with the information on the PacBio runs (here we just have one):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pacbio_runs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;flu_WSN&quot;</span><span class="p">],</span> <span class="s2">&quot;fastq&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;input_files/flu_WSN_ccs.fastq&quot;</span><span class="p">]}</span>
<span class="p">)</span>

<span class="n">pacbio_runs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>fastq</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>flu_WSN</td>
      <td>input_files/flu_WSN_ccs.fastq</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Create an <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.ccs.html#alignparse.ccs.Summaries">alignparse.ccs.Summaries</a> object:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ccs_summaries</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">ccs</span><span class="o">.</span><span class="n">Summaries</span><span class="p">(</span><span class="n">pacbio_runs</span><span class="p">,</span> <span class="n">report_col</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Statistics on the CCSs (length, number of subread passes, quality):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="k">for</span> <span class="n">stat</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">,</span> <span class="s2">&quot;passes&quot;</span><span class="p">,</span> <span class="s2">&quot;accuracy&quot;</span><span class="p">]:</span>
    <span class="k">if</span> <span class="n">ccs_summaries</span><span class="o">.</span><span class="n">has_stat</span><span class="p">(</span><span class="n">stat</span><span class="p">):</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">ccs_summaries</span><span class="o">.</span><span class="n">plot_ccs_stats</span><span class="p">(</span><span class="n">stat</span><span class="p">)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No </span><span class="si">{</span><span class="n">stat</span><span class="si">}</span><span class="s2"> statistics available.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Align-and-parse">
<h2>Align and parse<a class="headerlink" href="#Align-and-parse" title="Link to this heading">¶</a></h2>
<p>First, we create an <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.minimap2.html#alignparse.minimap2.Mapper">alignparse.minimap2.Mapper</a> to run <a class="reference external" href="https://github.com/lh3/minimap2">minimap2</a>, which is used for the alignments. We use <a class="reference external" href="https://github.com/lh3/minimap2">minimap2</a> options that are tailored for viruses with potentially large internal deletions (which are handled like spliced introns); these options are specified by
<a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.minimap2.html#alignparse.minimap2.OPTIONS_VIRUS_W_DEL">alignparse.minimap2.OPTIONS_VIRUS_W_DEL</a>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">mapper</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">minimap2</span><span class="o">.</span><span class="n">Mapper</span><span class="p">(</span><span class="n">alignparse</span><span class="o">.</span><span class="n">minimap2</span><span class="o">.</span><span class="n">OPTIONS_VIRUS_W_DEL</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Using `minimap2` </span><span class="si">{</span><span class="n">mapper</span><span class="o">.</span><span class="n">version</span><span class="si">}</span><span class="s2"> with these options:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="o">+</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mapper</span><span class="o">.</span><span class="n">options</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Using `minimap2` 2.22-r1101 with these options:
-xsplice:hq -un -C0 --splice-flank=no -M=1 --end-seed-pen=2 --end-bonus=2 --secondary=no --cs
</pre></div></div>
</div>
<p>Now use <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.align_and_parse">Targets.align_and_parse</a> to align and parse the CCSs. First, create the output directory:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">align_and_parse_outdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s2">&quot;flu_WSN_align_and_parse&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now do the alignments and parsing:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">readstats</span><span class="p">,</span> <span class="n">aligned</span><span class="p">,</span> <span class="n">filtered</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">align_and_parse</span><span class="p">(</span>
    <span class="n">df</span><span class="o">=</span><span class="n">pacbio_runs</span><span class="p">,</span>
    <span class="n">mapper</span><span class="o">=</span><span class="n">mapper</span><span class="p">,</span>
    <span class="n">outdir</span><span class="o">=</span><span class="n">align_and_parse_outdir</span><span class="p">,</span>
    <span class="n">name_col</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">,</span>
    <span class="n">queryfile_col</span><span class="o">=</span><span class="s2">&quot;fastq&quot;</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># overwrite any existing output</span>
    <span class="n">ncpus</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>  <span class="c1"># use all available CPUs</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>First, let’s look at the read alignment stats:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">readstats</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>category</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>flu_WSN</td>
      <td>filtered fluPB2</td>
      <td>15</td>
    </tr>
    <tr>
      <th>1</th>
      <td>flu_WSN</td>
      <td>aligned fluPB2</td>
      <td>50</td>
    </tr>
    <tr>
      <th>2</th>
      <td>flu_WSN</td>
      <td>filtered fluPB1</td>
      <td>34</td>
    </tr>
    <tr>
      <th>3</th>
      <td>flu_WSN</td>
      <td>aligned fluPB1</td>
      <td>165</td>
    </tr>
    <tr>
      <th>4</th>
      <td>flu_WSN</td>
      <td>filtered fluPA</td>
      <td>36</td>
    </tr>
    <tr>
      <th>5</th>
      <td>flu_WSN</td>
      <td>aligned fluPA</td>
      <td>173</td>
    </tr>
    <tr>
      <th>6</th>
      <td>flu_WSN</td>
      <td>filtered fluHA</td>
      <td>2</td>
    </tr>
    <tr>
      <th>7</th>
      <td>flu_WSN</td>
      <td>aligned fluHA</td>
      <td>8</td>
    </tr>
    <tr>
      <th>8</th>
      <td>flu_WSN</td>
      <td>filtered fluNP</td>
      <td>6</td>
    </tr>
    <tr>
      <th>9</th>
      <td>flu_WSN</td>
      <td>aligned fluNP</td>
      <td>23</td>
    </tr>
    <tr>
      <th>10</th>
      <td>flu_WSN</td>
      <td>filtered fluNA</td>
      <td>3</td>
    </tr>
    <tr>
      <th>11</th>
      <td>flu_WSN</td>
      <td>aligned fluNA</td>
      <td>18</td>
    </tr>
    <tr>
      <th>12</th>
      <td>flu_WSN</td>
      <td>filtered fluM1</td>
      <td>7</td>
    </tr>
    <tr>
      <th>13</th>
      <td>flu_WSN</td>
      <td>aligned fluM1</td>
      <td>14</td>
    </tr>
    <tr>
      <th>14</th>
      <td>flu_WSN</td>
      <td>filtered fluM2</td>
      <td>1</td>
    </tr>
    <tr>
      <th>15</th>
      <td>flu_WSN</td>
      <td>aligned fluM2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>16</th>
      <td>flu_WSN</td>
      <td>filtered fluNS1</td>
      <td>3</td>
    </tr>
    <tr>
      <th>17</th>
      <td>flu_WSN</td>
      <td>aligned fluNS1</td>
      <td>20</td>
    </tr>
    <tr>
      <th>18</th>
      <td>flu_WSN</td>
      <td>filtered fluNS2</td>
      <td>0</td>
    </tr>
    <tr>
      <th>19</th>
      <td>flu_WSN</td>
      <td>aligned fluNS2</td>
      <td>0</td>
    </tr>
    <tr>
      <th>20</th>
      <td>flu_WSN</td>
      <td>unmapped</td>
      <td>108</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Plot these stats:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">readstats</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">category</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span>
                <span class="n">x</span><span class="p">[</span><span class="s2">&quot;category&quot;</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;category&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">),</span>
            <span class="n">is_aligned</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;category&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;aligned&quot;</span><span class="p">),</span>
        <span class="p">),</span>
        <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s2">&quot;is_aligned&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~ name&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">0.3</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">readstats</span><span class="p">),</span> <span class="mf">2.5</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Now let’s look at the reason that reads that were filtered failed to align. Recall that <code class="docutils literal notranslate"><span class="pre">filtered</span></code> holds a data frame for each target:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">[:</span><span class="mi">1</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;First few lines of `filtered` for </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">filtered</span><span class="p">[</span><span class="n">target</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
First few lines of `filtered` for fluPB2:
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>query_name</th>
      <th>filter_reason</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/5308608/ccs</td>
      <td>termini5 clip5</td>
    </tr>
    <tr>
      <th>1</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/6554482/ccs</td>
      <td>polyA mutation_nt_count</td>
    </tr>
    <tr>
      <th>2</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/7143925/ccs</td>
      <td>polyA mutation_nt_count</td>
    </tr>
    <tr>
      <th>3</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/9437481/ccs</td>
      <td>final_mRNA_nts mutation_op_count</td>
    </tr>
    <tr>
      <th>4</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/9830886/ccs</td>
      <td>polyA mutation_nt_count</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Concatenate the information for all of the targets and plot reasons why mapped reads were filtered:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>
<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">filtered</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">gene</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;filter_reason&quot;</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="o">+</span> <span class="n">geom_bar</span><span class="p">()</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~ gene&quot;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
        <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
        <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Now let’s look at the first few aligned reads for the first few genes:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">[:</span><span class="mi">4</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">First few entries in `aligned` for </span><span class="si">{</span><span class="n">target</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">aligned</span><span class="p">[</span><span class="n">target</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

First few entries in `aligned` for fluPB2:
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>query_name</th>
      <th>query_clip5</th>
      <th>query_clip3</th>
      <th>sequenced_mRNA_mutations</th>
      <th>sequenced_mRNA_accuracy</th>
      <th>UMI_sequence</th>
      <th>UMI_accuracy</th>
      <th>cellbarcode_sequence</th>
      <th>cellbarcode_accuracy</th>
      <th>variant_tag_1_sequence</th>
      <th>variant_tag_2_sequence</th>
      <th>variant_tag_3_sequence</th>
      <th>variant_tag_4_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/5243086/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del279to1904 C2077A C2157A C2214A</td>
      <td>1.0</td>
      <td>CAATAGGTAC</td>
      <td>1.0</td>
      <td>AGTAGTCACTTCAAGC</td>
      <td>1.0</td>
      <td>G</td>
      <td>G</td>
      <td>C</td>
      <td>C</td>
    </tr>
    <tr>
      <th>1</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/5833039/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del117to2070</td>
      <td>1.0</td>
      <td>AGCGCTCTTT</td>
      <td>1.0</td>
      <td>ATAAGGTGACCTCTCA</td>
      <td>1.0</td>
      <td>A</td>
      <td>A</td>
      <td>T</td>
      <td>T</td>
    </tr>
    <tr>
      <th>2</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/5898400/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>G67T del152to1989</td>
      <td>1.0</td>
      <td>TGTCACTGGC</td>
      <td>1.0</td>
      <td>TCGGAGCTGGTCTTAG</td>
      <td>1.0</td>
      <td>G</td>
      <td>G</td>
      <td>C</td>
      <td>C</td>
    </tr>
    <tr>
      <th>3</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/6160599/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>G242T C430G del551to2282</td>
      <td>1.0</td>
      <td>CTCGGTCCGC</td>
      <td>1.0</td>
      <td>GTGCGCATGGAAATCA</td>
      <td>1.0</td>
      <td>G</td>
      <td>G</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>4</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/6161088/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del152to1989 G2057T G2118T</td>
      <td>1.0</td>
      <td>GGTTGGCGCT</td>
      <td>1.0</td>
      <td>CGGAGCTGGTCTTTAG</td>
      <td>1.0</td>
      <td>G</td>
      <td>G</td>
      <td>C</td>
      <td>C</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

First few entries in `aligned` for fluPB1:
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>query_name</th>
      <th>query_clip5</th>
      <th>query_clip3</th>
      <th>sequenced_mRNA_mutations</th>
      <th>sequenced_mRNA_accuracy</th>
      <th>UMI_sequence</th>
      <th>UMI_accuracy</th>
      <th>cellbarcode_sequence</th>
      <th>cellbarcode_accuracy</th>
      <th>variant_tag_1_sequence</th>
      <th>variant_tag_2_sequence</th>
      <th>variant_tag_3_sequence</th>
      <th>variant_tag_4_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4784702/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del171to2016 del2268to2268</td>
      <td>1.000000</td>
      <td>GCGTATCCTG</td>
      <td>1.000000</td>
      <td>CCTTCGTGATGATTGC</td>
      <td>1.000000</td>
      <td>T</td>
      <td>C</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>1</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4785077/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del180to2032</td>
      <td>1.000000</td>
      <td>GTGCATGTGT</td>
      <td>1.000000</td>
      <td>TTCAGCCTGTTCAAGC</td>
      <td>1.000000</td>
      <td>T</td>
      <td>C</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>2</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4915675/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del171to2016 del2070to2070</td>
      <td>1.000000</td>
      <td>CACAAACTAT</td>
      <td>1.000000</td>
      <td>GCACAGCTGACTGATC</td>
      <td>1.000000</td>
      <td>T</td>
      <td>C</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>3</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/5111911/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>ins397T C468T T1612C</td>
      <td>0.999384</td>
      <td>AGCGTGCCAC</td>
      <td>0.999999</td>
      <td>ATCCTGGCTGCGTCCA</td>
      <td>0.999998</td>
      <td>T</td>
      <td>C</td>
      <td>C</td>
      <td>T</td>
    </tr>
    <tr>
      <th>4</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/5112319/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del258to2001</td>
      <td>1.000000</td>
      <td>CAACGATGAC</td>
      <td>1.000000</td>
      <td>GGCCACAACGCGAATG</td>
      <td>1.000000</td>
      <td>T</td>
      <td>C</td>
      <td>C</td>
      <td>T</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

First few entries in `aligned` for fluPA:
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>query_name</th>
      <th>query_clip5</th>
      <th>query_clip3</th>
      <th>sequenced_mRNA_mutations</th>
      <th>sequenced_mRNA_accuracy</th>
      <th>UMI_sequence</th>
      <th>UMI_accuracy</th>
      <th>cellbarcode_sequence</th>
      <th>cellbarcode_accuracy</th>
      <th>variant_tag_1_sequence</th>
      <th>variant_tag_2_sequence</th>
      <th>variant_tag_3_sequence</th>
      <th>variant_tag_4_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4587763/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del117to1927 C2159A</td>
      <td>1.000000</td>
      <td>AAACTAGTGG</td>
      <td>1.0</td>
      <td>AAGACATGATCAGTGT</td>
      <td>1.0</td>
      <td>T</td>
      <td></td>
      <td>T</td>
      <td>C</td>
    </tr>
    <tr>
      <th>1</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4653834/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del179to1839</td>
      <td>1.000000</td>
      <td>AGCACACATA</td>
      <td>1.0</td>
      <td>TCCCATTGAAAACACA</td>
      <td>1.0</td>
      <td>C</td>
      <td>T</td>
      <td>A</td>
      <td>T</td>
    </tr>
    <tr>
      <th>2</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4719236/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>C42A del179to1839</td>
      <td>1.000000</td>
      <td>ACTGAAACAA</td>
      <td>1.0</td>
      <td>TCCCGTTGAAAACACA</td>
      <td>1.0</td>
      <td>C</td>
      <td>T</td>
      <td>A</td>
      <td>T</td>
    </tr>
    <tr>
      <th>3</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4784775/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del66to177 del265to1996 C2028T</td>
      <td>0.999981</td>
      <td>AATAGTTCCC</td>
      <td>1.0</td>
      <td>GGCAAACGAGACGCAA</td>
      <td>1.0</td>
      <td></td>
      <td></td>
      <td>A</td>
      <td>T</td>
    </tr>
    <tr>
      <th>4</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4850043/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>A64G ins402A del1063to1067 del1073to2180</td>
      <td>0.999704</td>
      <td>CATCATATTG</td>
      <td>1.0</td>
      <td>TGTGGTCCTACTGATC</td>
      <td>1.0</td>
      <td>T</td>
      <td>C</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

First few entries in `aligned` for fluHA:
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>query_name</th>
      <th>query_clip5</th>
      <th>query_clip3</th>
      <th>sequenced_mRNA_mutations</th>
      <th>sequenced_mRNA_accuracy</th>
      <th>UMI_sequence</th>
      <th>UMI_accuracy</th>
      <th>cellbarcode_sequence</th>
      <th>cellbarcode_accuracy</th>
      <th>variant_tag_1_sequence</th>
      <th>variant_tag_2_sequence</th>
      <th>variant_tag_3_sequence</th>
      <th>variant_tag_4_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/4522378/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>G397A ins1145A del1251to1722</td>
      <td>1.000000</td>
      <td>TAGACCCTTT</td>
      <td>1.00000</td>
      <td>GAGTACACTATCGCGC</td>
      <td>1.0</td>
      <td>C</td>
      <td>G</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th>1</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/6947580/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>del108to1459</td>
      <td>1.000000</td>
      <td>CTTTGGGGGG</td>
      <td>0.99998</td>
      <td>GTAGGTTTGTCACTAG</td>
      <td>1.0</td>
      <td></td>
      <td></td>
      <td>G</td>
      <td>T</td>
    </tr>
    <tr>
      <th>2</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/7144230/ccs</td>
      <td>2</td>
      <td>0</td>
      <td>C260A C262A ins288G del471to471 C1052A ins1450T</td>
      <td>0.999627</td>
      <td>CGTACGGCAG</td>
      <td>1.00000</td>
      <td>AAACTGGCTAGGAGCT</td>
      <td>1.0</td>
      <td>A</td>
      <td>A</td>
      <td>A</td>
      <td>C</td>
    </tr>
    <tr>
      <th>3</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/7340424/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>G213A ins663A G1715A del1718to1722</td>
      <td>0.999847</td>
      <td>GCTTGCCGGC</td>
      <td>1.00000</td>
      <td>GGTCACTCTGCGTTAT</td>
      <td>1.0</td>
      <td>A</td>
      <td>A</td>
      <td>A</td>
      <td>C</td>
    </tr>
    <tr>
      <th>4</th>
      <td>flu_WSN</td>
      <td>m54228_171201_172448/8520141/ccs</td>
      <td>0</td>
      <td>0</td>
      <td>G9T C153A del771to771 A1253C del1256to1722</td>
      <td>0.999810</td>
      <td>GTGCCAAGGA</td>
      <td>1.00000</td>
      <td>GAAAGGCTGGTACGAG</td>
      <td>1.0</td>
      <td>C</td>
      <td>G</td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo" />
    
    <h1 class="logo logo-name">alignparse</h1>
    
  </a>
</p>



<p class="blurb">Align sequences and then parse features.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=alignparse&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">alignparse documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="recA_DMS.html">RecA deep mutational scanning libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="lasv_pilot.html">Lassa virus glycoprotein PacBio sequencing</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Single-cell virus sequencing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="alignparse.html">alignparse package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">Package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="examples.html">Examples</a><ul>
      <li>Previous: <a href="lasv_pilot.html" title="previous chapter">Lassa virus glycoprotein PacBio sequencing</a></li>
      <li>Next: <a href="alignparse.html" title="next chapter">alignparse package</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2019--2025.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/flu_virus_seq_example.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/alignparse" class="github">
        <img src="_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>