<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Lassa virus glycoprotein PacBio sequencing &#8212; alignparse 0.7.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=877b1dcb" />
    <link rel="stylesheet" type="text/css" href="_static/plot_directive.css?v=7f9a90b1" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />
    <script src="_static/documentation_options.js?v=fe7df9b0"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Single-cell virus sequencing" href="flu_virus_seq_example.html" />
    <link rel="prev" title="RecA deep mutational scanning libraries" href="recA_DMS.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="Lassa-virus-glycoprotein-PacBio-sequencing">
<h1>Lassa virus glycoprotein PacBio sequencing<a class="headerlink" href="#Lassa-virus-glycoprotein-PacBio-sequencing" title="Link to this heading">¶</a></h1>
<p>This example shows how to use <a class="reference external" href="https://jbloomlab.github.io/alignparse/">alignparse</a> to process circular consensus sequences that map to multiple different targets. Here, our two targets are a wildtype and a codon optimized sequence for the Lassa virus (LASV) glycoprotein from the Josiah strain. For such targets we do not expect large internal deletions, so we use alignment settings optimized for codon-level mutations as these will be the settings used when analyzing mutant LASV GP sequences
in later experiments.</p>
<p>Here we analyze a snippet of the full data set of circular consensus sequences so that the example is small and fast.</p>
<p>In the other included example notebooks (<a class="reference external" href="https://jbloomlab.github.io/alignparse/recA_DMS.html">RecA deep mutational scanning libraries</a> and <a class="reference external" href="https://jbloomlab.github.io/alignparse/flu_virus_seq_example.html">Single-cell virus sequencing</a>), we align and parse the PacBio reads using the single <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.align_and_parse">align_and_parse</a> function. Here we use separate functions to do the aligning and parsing
steps. This illustrates an additional use case and shows how this package could be used to parse alignments generated elsewhere as long as they have a <a class="reference external" href="https://lh3.github.io/minimap2/minimap2.html#10">cs tag</a> and are in the SAM file format.</p>
<section id="Set-up-for-analysis">
<h2>Set up for analysis<a class="headerlink" href="#Set-up-for-analysis" title="Link to this heading">¶</a></h2>
<p>Import necessary Python modules:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tempfile</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">Bio.SeqIO</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">plotnine</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pysam</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.ccs</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.minimap2</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.targets</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.cs_tag</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">alignparse.consensus</span>
</pre></div>
</div>
</div>
<p>Suppress warnings that clutter output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Directory for output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outdir</span> <span class="o">=</span> <span class="s2">&quot;./output_files/&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Color palette for plots:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">CBPALETTE</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;#999999&quot;</span><span class="p">,</span> <span class="s2">&quot;#E69F00&quot;</span><span class="p">,</span> <span class="s2">&quot;#56B4E9&quot;</span><span class="p">,</span> <span class="s2">&quot;#009E73&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Target-amplicons">
<h2>Target amplicons<a class="headerlink" href="#Target-amplicons" title="Link to this heading">¶</a></h2>
<p>We have performed sequencing of several LASV GP amplicons that include the glycoprotein sequence along with a PacBio index and several other features. Here we analyze reads mapping to two of these amplicons. The amplicons are defined in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/samplerecord/">Genbank Flat File format</a>. If there are multiple targets, they can all be defined in a single Genbank file (as we did in the <a class="reference external" href="https://jbloomlab.github.io/alignparse/flu_virus_seq_example.html">Single-cell virus
sequencing</a> example) or each target can be defined in its own Genbank file (as we’ve done here).</p>
<p>First, let’s just look at the files:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">target_file_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;LASV_Josiah_WT&quot;</span><span class="p">,</span> <span class="s2">&quot;LASV_Josiah_OPT&quot;</span><span class="p">]</span>

<span class="n">targetfiles</span> <span class="o">=</span> <span class="p">[</span>
    <span class="sa">f</span><span class="s2">&quot;input_files/</span><span class="si">{</span><span class="n">target_file_name</span><span class="si">}</span><span class="s2">.gb&quot;</span> <span class="k">for</span> <span class="n">target_file_name</span> <span class="ow">in</span> <span class="n">target_file_names</span>
<span class="p">]</span>

<span class="k">for</span> <span class="n">targetfile</span> <span class="ow">in</span> <span class="n">targetfiles</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">targetfile</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LOCUS       LASV_Josiah_WT          1730 bp ds-DNA     linear       14-JUN-2019
DEFINITION  .
ACCESSION
VERSION
SOURCE      Kate Crawford
  ORGANISM  .
COMMENT     PacBio amplicon for LASV Josiah WT sequence
FEATURES             Location/Qualifiers
     T2A             85..147
                     /label=&#34;T2A&#34;
     WPRE            1639..1730
                     /label=&#34;WPRE&#34;
     ZsGreen         15..84
                     /label=&#34;ZsGreen&#34;
     termini3        1639..1730
                     /label=&#34;3&#39;Termini&#34;
     index           9..14
                     /label=&#34;index&#34;
     leader5         1..8
                     /label=&#34;5&#39; leader&#34;
     termini5        1..147
                     /label=&#34;5&#39;Termini&#34;
     variant_tag5    34..34
                     /variant_1=T
                     /variant_2=C
                     /label=&#34;5&#39;VariantTag&#34;
     variant_tag3    1702..1702
                     /variant_1=G
                     /variant_2=A
                     /label=&#34;3&#39;VariantTag&#34;
     spacer          1624..1638
                     /label=&#34;3&#39;Spacer&#34;
     gene            148..1623
                     /label=&#34;LASV_Josiah_WT&#34;

ORIGIN
        1 GACTGATANN NNNNcagcga cgccaagaac cagYagtggc acctgaccga gcacgccatc
       61 gcctccggcT CCGCCTTGCC CGCTGGATCC GGCGAGGGCA GAGGAAGTCT GCTAACATGC
      121 GGTGACGTCG AGGAGAATCC TGGCCCAATG GGACAAATAG TGACATTCTT CCAGGAAGTG
      181 CCTCATGTAA TAGAAGAGGT GATGAACATT GTTCTCATTG CACTGTCTGT ACTAGCAGTG
      241 CTGAAAGGTC TGTACAATTT TGCAACGTGT GGCCTTGTTG GTTTGGTCAC TTTCCTCCTG
      301 TTGTGTGGTA GGTCTTGCAC AACCAGTCTT TATAAAGGGG TTTATGAGCT TCAGACTCTG
      361 GAACTAAACA TGGAGACACT CAATATGACC ATGCCTCTCT CCTGCACAAA GAACAACAGT
      421 CATCATTATA TAATGGTGGG CAATGAGACA GGACTAGAAC TGACCTTGAC CAACACGAGC
      481 ATTATTAATC ACAAATTTTG CAATCTGTCT GATGCCCACA AAAAGAACCT CTATGACCAC
      541 GCTCTTATGA GCATAATCTC AACTTtccac ttgtccatcc ccaacTTCAA TCAGTATGAG
      601 GCAATGAGCT GCGATTTTAA TGGGGGAAAG ATTAGTGTGC AGTACAACCT GAGTCACAGC
      661 TATGCTGGGG ATGCAGCCAA CCATTGTGGT ACTGTTGCAA ATGGTGTGTT ACAGACTTTT
      721 ATGAGGATGG CTTGGGGTGG GAGCTACATT GCTCTTGACT CAGGCCGTGG CAACTGGGAC
      781 TGTATTATGA CTAGTTATCA ATATCTGATA ATCCAAAATA CAACCTGGGA AGATCACTGC
      841 CAATTCTCGA GACCATCTCC CATCGGTTAT CTCGGGCTCC TCTCACAAAG GACTAGAGAT
      901 ATTTATATTA GTAGAAGATT GCTAGGCACA TTCACATGGA CACTGTCAGA TTCTGAAGGT
      961 AAAGACACAC CAGGGGGATA TTGTCTGACC AGGTGGATGC TAATTGAGGC TGAACTAAAA
     1021 TGCTTCGGGA ACACAGCTGT GGCAAAATGT AATGAGAAGC ATGATGAgga attttgtgac
     1081 atgctgaggc TGTTTGACTT CAACAAACAA GCCATTCAAA GGTTGAAAGC TGAAGCACAA
     1141 ATGAGCATTC AGTTGATCAA CAAAGCAGTA AATGCTTTGA TAAATGACCA ACTTATAATG
     1201 AAGAACCATC TACGGGACAT CATGGGAATT CCATACTGTA ATTACAGCAA GTATTGGTAC
     1261 CTCAACCACA CAACTACTGG GAGAACATCA CTGCCCAAAT GTTGGCTTGT ATCAAATGGT
     1321 TCATACTTGA ACGAGACCCA CTTTTCTGAT GATATTGAAC AACAAGCTGA CAATATGATC
     1381 ACTGAGATGT TACAGAAGGA GTATATGGAG AGGCAGGGGA AGACACCATT GGGTCTAGTT
     1441 GACCTCTTTG TGTTCAGCAC AAGTTTCTAT CTTATTAGCA TCTTCCTTCA CCTAGTCAAA
     1501 ATACCAACTC ATAGGCATAT TGTAGGCAAG TCGTGTCCCA AACCTCACAG ATTGAATCAT
     1561 ATGGGCATTT GTTCCTGTGG ACTCTACAAA CAGCCTGGTG TGCCTGTGAA ATGGAAGAGA
     1621 TGAGCTAGCT AAACGCGTTG ATCCtaatca acctctggat tacaaaattt gtgaaagatt
     1681 gactggtatt cttaactatg tRgctccttt tacgctatgt ggatacgctg
//

LOCUS       LASV_Josiah_OPT          1730 bp ds-DNA     linear       14-JUN-2019
DEFINITION  .
ACCESSION
VERSION
SOURCE      Kate Crawford
  ORGANISM  .
COMMENT     PacBio amplicon for LASV Josiah OPT sequence
FEATURES             Location/Qualifiers
     T2A             85..147
                     /label=&#34;T2A&#34;
     WPRE            1639..1730
                     /label=&#34;WPRE&#34;
     ZsGreen         15..84
                     /label=&#34;ZsGreen&#34;
     termini3        1639..1730
                     /label=&#34;3&#39;Termini&#34;
     index           9..14
                     /label=&#34;index&#34;
     leader5         1..8
                     /label=&#34;5&#39; leader&#34;
     termini5        1..147
                     /label=&#34;5&#39;Termini&#34;
     variant_tag5     34..34
                     /variant_1=T
                     /variant_2=C
                     /label=&#34;5&#39;VariantTag&#34;
     variant_tag3     1702..1702
                     /variant_1=G
                     /variant_2=A
                     /label=&#34;3&#39;VariantTag&#34;
     spacer          1624..1638
                     /label=&#34;3&#39;Spacer&#34;
     gene            148..1623
                     /label=&#34;LASV_Josiah_OPT&#34;

ORIGIN
        1 GACTGATANN NNNNcagcga cgccaagaac cagYagtggc acctgaccga gcacgccatc
       61 gcctccggcT CCGCCTTGCC CGCTGGATCC GGCGAGGGCA GAGGAAGTCT GCTAACATGC
      121 GGTGACGTCG AGGAGAATCC TGGCCCAATG GGCCAGATCG TGACCTTCTT CCAAGAAGTG
      181 CCTCATGTGA TTGAGGAGGT GATGAATATC GTGCTGATCG CTTTAAGCGT GCTGGCCGTT
      241 CTTAAGGGCC TCTATAACTT CGCCACTTGT GGTTTAGTCG GACTGGTGAC ATTTCTGCTG
      301 CTGTGTGGCA GATCTTGTAC CACATCTTTA TACAAGGGCG TGTACGAGCT GCAGACTTTA
      361 GAACTGAACA TGGAGACTTT AAACATGACC ATGCCTTTAA GCTGTACCAA GAACAATAGC
      421 CACCACTACA TCATGGTGGG CAACGAGACC GGTTTAGAAC TGACACTCAC CAACACCAGC
      481 ATTATCAACC ATAAGTTCTG CAACCTCTCC GACGCTCACA AGAAGAATTT ATACGACCAC
      541 GCTTTAATGA GCATCATCTC CACCTTCCAT CTCTCCATTC CTAATttcaa ccagtacgag
      601 gccatgAGCT GCGACTTTAA CGGCGGCAAG ATCTCCGTGC AGTACAATTT ATCCCATAGC
      661 TACGCCGGCG ATGCCGCCAA TCACTGCGGA ACCGTGGCCA ACGGCGTGCT GCAGACATTC
      721 ATGAGGATGG CTTGGGGCGG CTCCTATATC GCTTTAGACT CCGGCAGAGG AAACTGGGAC
      781 TGTATCATGA CCAGCTACCA ATATTTAATC ATTCAGAACA CCACATGGGA GGACCACTGC
      841 CAATTCTCTC GTCCCTCTCC TATCGGCTAT CTGGGACTGC TGTCCCAGAG GACCAGAGAC
      901 ATCTACATCT CTCGTAGGCT GCTGGGCACA TTCACTTGGA CTTTAAGCGA CAGCGAAGGC
      961 AAAGATACTC CCGGTGGCTA CTGTTTAACA AGATGGATGC TGATCGAGGC CGAGCTCAAG
     1021 TGCTTCGGAA ATACCGCCGT GGCCAAATGC AACGAGAAAC ACGACGAGGA GTTCTGCGAC
     1081 ATGCTGAGGC TCTTCGACTT CAacaagcaa gccattcaga ggcTGAAGGC CGAAGCCCAG
     1141 ATGTCCATCC AGCTGATTAA TAAGGCCGTG AATGCCCTCA TTAACGACCA GCTGATCATG
     1201 AAGAACCATT TAAGGGACAT CATGGGCATC CCTTATTGCA ACTACAGCAA ATACTGGTAT
     1261 TTAAATCATA CCACCACCGG TCGTACATCC TTACCTAAGT GCTGGCTGGT CAGCAATGGC
     1321 TCCTATTTAA ACGAGACACA CTTCTCCGAC GACATCGAGC AGCAAGCCGA CAACATGATC
     1381 ACCGAAATGC TCCAGAAGGA GTACATGGAG AGGCAAGGTA AGACTCCTCT GGGTTTAGTG
     1441 GATTTATTCG TCTTCAGCAC CTCCTTCTAT TTAATCTCCA TCTTTCTTCA TCTGGTGAAG
     1501 ATTCCTACCC ACAGACACAT TGTGGGCAAG AGCTGTCCTA AGCCTCATAG ACTGAACCAC
     1561 ATGGGCATCT GTAGCTGCGG TTTATATAAA CAGCCCGGTG TTCCCGTTAA GTGGAAGAGG
     1621 TGAGCTAGCT AAACGCGTTG ATCCtaatca acctctggat tacaaaattt gtgaaagatt
     1681 gactggtatt cttaactatg tRgctccttt tacgctatgt ggatacgctg
//

</pre></div></div>
</div>
<p>Along with the Genbank files giving the sequences of the amplicons, we have a YAML file specifying how to filter and parse alignments to these amplicons.</p>
<p>Below is the text of the YAML file.</p>
<p>Additional information about these filters can be found in the <a class="reference external" href="https://jbloomlab.github.io/alignparse/recA_DMS.html">RecA deep mutational scanning libraries</a> example notebook or the <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">Targets</a> documentation.</p>
<p>A filter setting of <code class="docutils literal notranslate"><span class="pre">null</span></code> indicates this filter is not applied. When filters are missing for a feature, they are automatically set to zero.</p>
<p>Here we filter the <code class="docutils literal notranslate"><span class="pre">gene</span></code> based on <code class="docutils literal notranslate"><span class="pre">mutation_op_counts</span></code> by setting the <code class="docutils literal notranslate"><span class="pre">mutation_nt_counts</span></code> filter for the <code class="docutils literal notranslate"><span class="pre">gene</span></code> to <code class="docutils literal notranslate"><span class="pre">null</span></code>. Although we do not expect these sequences to have large indels, we want to confirm this. Filtering on mutation “operations” allows us to retain sequences with large indels by only filtering on the number of indels, not the number of nucleotides inserted or deleted. Overall, this example uses very loose filters to allow us to do further analyses of the types of
mutations that are arising in these samples.</p>
<p>The YAML file also specifies what information is parsed from alignments that are not filtered out. As you can see, for some features we parse the mutations or the full sequence of the feature, along with the accuracy of that feature in the sequencing query (computed from the Q-values).</p>
<p>As seen below, we can use YAML syntax to apply one set of filters to multiple targets. Here, we apply the same filters to both targets, but this is not necessary.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">lasv_parse_specs_file</span> <span class="o">=</span> <span class="s2">&quot;input_files/lasv_feature_parse_specs.yaml&quot;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">lasv_parse_specs_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_WT: &amp;LASV_target_parse_specs
  query_clip5: 10
  query_clip3: 10
  termini5:
    filter:
      clip5: 10
      mutation_nt_count: 5
      mutation_op_count: null
  termini3:
    filter:
      clip3: 10
      mutation_nt_count: 5
      mutation_op_count: null
  gene:
    filter:
      mutation_nt_count: null
      mutation_op_count: 30
    return: [mutations, accuracy]
  spacer:
    filter:
      mutation_nt_count: 1
      mutation_op_count: null
  index:
    return: sequence
  variant_tag5:
    return: sequence
  variant_tag3:
    return: sequence

LASV_Josiah_OPT: *LASV_target_parse_specs

</pre></div></div>
</div>
<p>Read the amplicons in <code class="docutils literal notranslate"><span class="pre">feature_parse_specs</span></code> into a <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">Targets</a> object, specifying the features that we require the target to contain. The <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">Targets</a> in this example have more features specified in their Genbank files than we want to parse, so we set <code class="docutils literal notranslate"><span class="pre">allow_extra_features</span></code> to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">targets</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">targets</span><span class="o">.</span><span class="n">Targets</span><span class="p">(</span>
    <span class="n">seqsfile</span><span class="o">=</span><span class="n">targetfiles</span><span class="p">,</span>
    <span class="n">feature_parse_specs</span><span class="o">=</span><span class="n">lasv_parse_specs_file</span><span class="p">,</span>
    <span class="n">allow_extra_features</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>When we look at the <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.feature_parse_specs">targets.feature_parse_specs</a>, we now see that the previously unspecified specs have been filled in with the defaults.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">targets</span><span class="o">.</span><span class="n">feature_parse_specs</span><span class="p">(</span><span class="s2">&quot;yaml&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_WT: &amp;id001
  query_clip5: 10
  query_clip3: 10
  termini5:
    filter:
      clip5: 10
      mutation_nt_count: 5
      mutation_op_count: null
      clip3: 0
    return: []
  termini3:
    filter:
      clip3: 10
      mutation_nt_count: 5
      mutation_op_count: null
      clip5: 0
    return: []
  gene:
    filter:
      mutation_nt_count: null
      mutation_op_count: 30
      clip5: 0
      clip3: 0
    return:
    - mutations
    - accuracy
  spacer:
    filter:
      mutation_nt_count: 1
      mutation_op_count: null
      clip5: 0
      clip3: 0
    return: []
  index:
    return:
    - sequence
    filter:
      clip5: 0
      clip3: 0
      mutation_nt_count: 0
      mutation_op_count: 0
  variant_tag5:
    return:
    - sequence
    filter:
      clip5: 0
      clip3: 0
      mutation_nt_count: 0
      mutation_op_count: 0
  variant_tag3:
    return:
    - sequence
    filter:
      clip5: 0
      clip3: 0
      mutation_nt_count: 0
      mutation_op_count: 0
LASV_Josiah_OPT: *id001

</pre></div></div>
</div>
<p>We can also plot the <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">Targets</a>. All features specified in the targets’ Genbank files will be annotated, even if they are not in <code class="docutils literal notranslate"><span class="pre">feature_parse_specs</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax_width</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/lasv_pilot_18_0.png" src="_images/lasv_pilot_18_0.png" />
</div>
</div>
<p>Note that if needed, it is also possible to get a <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">Targets</a> object for just some of the sequences specified in <code class="docutils literal notranslate"><span class="pre">seqsfile</span></code> or <code class="docutils literal notranslate"><span class="pre">feature_parse_specs</span></code>. This is most commonly useful when <code class="docutils literal notranslate"><span class="pre">seqsfile</span></code> contains additional sequences that are not of interest. Below we illustrate how to do this by:</p>
<ul class="simple">
<li><p>Setting <code class="docutils literal notranslate"><span class="pre">select_target_names</span></code> to only keep the <em>LASV_Josiah_WT</em> sequence in <code class="docutils literal notranslate"><span class="pre">seqsfile</span></code></p></li>
<li><p>Setting <code class="docutils literal notranslate"><span class="pre">ingore_feature_parse_specs</span></code> to ignore the other targets (in this case, <em>LASV_Josiah_OPT</em>) in <code class="docutils literal notranslate"><span class="pre">feature_parse_specs</span></code></p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">targets_subset</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">targets</span><span class="o">.</span><span class="n">Targets</span><span class="p">(</span>
    <span class="n">seqsfile</span><span class="o">=</span><span class="n">targetfiles</span><span class="p">,</span>
    <span class="n">feature_parse_specs</span><span class="o">=</span><span class="n">lasv_parse_specs_file</span><span class="p">,</span>
    <span class="n">allow_extra_features</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">select_target_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;LASV_Josiah_WT&quot;</span><span class="p">],</span>
    <span class="n">ignore_feature_parse_specs_keys</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;LASV_Josiah_OPT&quot;</span><span class="p">],</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Here are the names of the retained targets: </span><span class="si">{</span><span class="n">targets_subset</span><span class="o">.</span><span class="n">target_names</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Here are the names of the retained targets: [&#39;LASV_Josiah_WT&#39;]
</pre></div></div>
</div>
</section>
<section id="PacBio-CCSs">
<h2>PacBio CCSs<a class="headerlink" href="#PacBio-CCSs" title="Link to this heading">¶</a></h2>
<p>We will align PacBio circular consensus sequences (CCSs) to the target. First, we want to look at the CCSs. A FASTQ file with these CCSs along with an associated report file were generated using the PacBio <code class="docutils literal notranslate"><span class="pre">ccs</span></code> program (see <a class="reference external" href="https://github.com/PacificBiosciences/ccs">here</a> for details on <code class="docutils literal notranslate"><span class="pre">ccs</span></code>) using commands like the following (generates report file and BAM of CCSs):</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>ccs --minLength 50 --maxLength 5000 \
    --minPasses 3  --minPredictedAccuracy 0.999 \
    --reportFile lasv_pilot_report.txt \
    --polish --numThreads 16 \
    lasv_pilot_subreads.bam lasv_pilot_ccs.bam
</pre></div>
</div>
<p>The BAM file was then converted to a FASTQ file using <a class="reference external" href="http://www.htslib.org/">samtools</a> with flags to retain the number of passes (<code class="docutils literal notranslate"><span class="pre">np</span></code>) and read quality (<code class="docutils literal notranslate"><span class="pre">rq</span></code>). Here we convert the BAM file to a gzipped FASTQ file to demonstrate the use of compressed files:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>samtools bam2fq -T np,rq lasv_pilot_ccs.bam | gzip &gt; lasv_pilot_ccs.fastq.gz
</pre></div>
</div>
<p>Here is a data frame with the resulting FASTQ and BAM files:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">run_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;lasv_pilot&quot;</span><span class="p">]</span>
<span class="n">ccs_dir</span> <span class="o">=</span> <span class="s2">&quot;input_files&quot;</span>
<span class="n">file_name</span> <span class="o">=</span> <span class="s2">&quot;lasv_example&quot;</span>

<span class="n">pacbio_runs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">run_names</span><span class="p">,</span>
        <span class="s2">&quot;report&quot;</span><span class="p">:</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ccs_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">_report.txt&quot;</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">run_names</span><span class="p">],</span>
        <span class="s2">&quot;fastq&quot;</span><span class="p">:</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ccs_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">_ccs.fastq.gz&quot;</span><span class="p">],</span>
    <span class="p">}</span>
<span class="p">)</span>

<span class="n">pacbio_runs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>report</th>
      <th>fastq</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lasv_pilot</td>
      <td>input_files/lasv_pilot_report.txt</td>
      <td>input_files/lasv_example_ccs.fastq.gz</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We create a <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.ccs.html#alignparse.ccs.Summaries">Summaries</a> object for these CCSs:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ccs_summaries</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">ccs</span><span class="o">.</span><span class="n">Summaries</span><span class="p">(</span><span class="n">pacbio_runs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot how many ZMWs yielded CCSs:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">p</span> <span class="o">=</span> <span class="n">ccs_summaries</span><span class="o">.</span><span class="n">plot_zmw_stats</span><span class="p">()</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ccs_summaries</span><span class="o">.</span><span class="n">zmw_stats</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>status</th>
      <th>number</th>
      <th>fraction</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lasv_pilot</td>
      <td>Success -- CCS generated</td>
      <td>250</td>
      <td>0.4996</td>
    </tr>
    <tr>
      <th>1</th>
      <td>lasv_pilot</td>
      <td>Failed -- Not enough full passes</td>
      <td>157</td>
      <td>0.3144</td>
    </tr>
    <tr>
      <th>2</th>
      <td>lasv_pilot</td>
      <td>Failed -- CCS below minimum predicted accuracy</td>
      <td>86</td>
      <td>0.1720</td>
    </tr>
    <tr>
      <th>3</th>
      <td>lasv_pilot</td>
      <td>Failed -- No usable subreads</td>
      <td>6</td>
      <td>0.0137</td>
    </tr>
    <tr>
      <th>4</th>
      <td>lasv_pilot</td>
      <td>Failed -- Other reason</td>
      <td>0</td>
      <td>0.0000</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Statistics on the CCSs (length, number of subread passes, accuracy):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="k">for</span> <span class="n">stat</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">,</span> <span class="s2">&quot;passes&quot;</span><span class="p">,</span> <span class="s2">&quot;accuracy&quot;</span><span class="p">]:</span>
    <span class="k">if</span> <span class="n">ccs_summaries</span><span class="o">.</span><span class="n">has_stat</span><span class="p">(</span><span class="n">stat</span><span class="p">):</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">ccs_summaries</span><span class="o">.</span><span class="n">plot_ccs_stats</span><span class="p">(</span><span class="n">stat</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No information available on CCS </span><span class="si">{</span><span class="n">stat</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Align-CCSs-to-target">
<h2>Align CCSs to target<a class="headerlink" href="#Align-CCSs-to-target" title="Link to this heading">¶</a></h2>
<p>Now we use <a class="reference external" href="https://github.com/lh3/minimap2">minimap2</a> to align the CCSs to the target.</p>
<p>First, we create a <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.minimap2.html#alignparse.minimap2.Mapper">Mapper</a> object to run <a class="reference external" href="https://github.com/lh3/minimap2">minimap2</a>, using the options for codon-level deep mutational scanning (specified by <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.minimap2.html#alignparse.minimap2.OPTIONS_CODON_DMS">alignparse.minimap2.OPTIONS_CODON_DMS</a>):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">mapper</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">minimap2</span><span class="o">.</span><span class="n">Mapper</span><span class="p">(</span><span class="n">alignparse</span><span class="o">.</span><span class="n">minimap2</span><span class="o">.</span><span class="n">OPTIONS_CODON_DMS</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Using `minimap2` </span><span class="si">{</span><span class="n">mapper</span><span class="o">.</span><span class="n">version</span><span class="si">}</span><span class="s2"> with these options:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="o">+</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mapper</span><span class="o">.</span><span class="n">options</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Using `minimap2` 2.22-r1101 with these options:
-A2 -B4 -O12 -E2 --end-bonus=13 --secondary=no --cs
</pre></div></div>
</div>
<p>Now use this mapper to do the alignments to a SAM file. First, add the names of the desired alignment files to our data frame:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pacbio_runs</span> <span class="o">=</span> <span class="n">pacbio_runs</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">alignments</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">outdir</span> <span class="o">+</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_alignments.sam&quot;</span>
<span class="p">)</span>

<span class="n">pacbio_runs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>report</th>
      <th>fastq</th>
      <th>alignments</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>lasv_pilot</td>
      <td>input_files/lasv_pilot_report.txt</td>
      <td>input_files/lasv_example_ccs.fastq.gz</td>
      <td>./output_files/lasv_pilot_alignments.sam</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Now we run <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.align">targets.align</a> using the mapper to actually align the FASTQ queries to the target:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">pacbio_runs</span><span class="o">.</span><span class="n">itertuples</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Aligning </span><span class="si">{</span><span class="n">tup</span><span class="o">.</span><span class="n">fastq</span><span class="si">}</span><span class="s2"> to create </span><span class="si">{</span><span class="n">tup</span><span class="o">.</span><span class="n">alignments</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
    <span class="n">targets</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">queryfile</span><span class="o">=</span><span class="n">tup</span><span class="o">.</span><span class="n">fastq</span><span class="p">,</span> <span class="n">alignmentfile</span><span class="o">=</span><span class="n">tup</span><span class="o">.</span><span class="n">alignments</span><span class="p">,</span> <span class="n">mapper</span><span class="o">=</span><span class="n">mapper</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Aligning input_files/lasv_example_ccs.fastq.gz to create ./output_files/lasv_pilot_alignments.sam...
</pre></div></div>
</div>
<p>These SAM files now contain the alignments along with the <a class="reference external" href="https://github.com/lh3/minimap2#cs">cs tag</a>.</p>
<p>An example <a class="reference external" href="https://github.com/lh3/minimap2#cs">cs tag</a> is:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">fname</span> <span class="ow">in</span> <span class="n">pacbio_runs</span><span class="p">[</span><span class="s2">&quot;alignments&quot;</span><span class="p">][:</span><span class="mi">1</span><span class="p">]:</span>
    <span class="k">with</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">a</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;First alignment in </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s2"> has `cs` tag:</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">a</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s2">&quot;cs&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
First alignment in ./output_files/lasv_pilot_alignments.sam has `cs` tag:
:8*ng*na*ng*na*nc*ng:19*nc:1667*na:28
</pre></div></div>
</div>
</section>
<section id="Parse-the-alignments">
<h2>Parse the alignments<a class="headerlink" href="#Parse-the-alignments" title="Link to this heading">¶</a></h2>
<p>Now we use <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.parse_alignment">Targets.parse_alignments</a> to parse the SAM files to get the information we specified for return. This function returns a data frame (<code class="docutils literal notranslate"><span class="pre">readstats</span></code>) on the overall parsing stats, plus dicts keyed by the names of each target in <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets">Targets</a> giving data frames of the aligned and filtered
reads. Here we return the <code class="docutils literal notranslate"><span class="pre">aligned</span></code> and <code class="docutils literal notranslate"><span class="pre">filtered</span></code> reads as dictionaries of data frames.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">aligned</span></code> and <code class="docutils literal notranslate"><span class="pre">filtered</span></code> read information can instead be returned as CSV files containing these data frames by setting the <code class="docutils literal notranslate"><span class="pre">to_csv</span></code> argument to <code class="docutils literal notranslate"><span class="pre">True</span></code>. Note: When using the combined <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.align_and_parse">Targets.align_and_parse</a> function as in the other example notebooks (<a class="reference external" href="https://jbloomlab.github.io/alignparse/recA_DMS.html">RecA DMS libraries</a> and <a class="reference external" href="https://jbloomlab.github.io/alignparse/flu_virus_seq_example.html">Single-cell virus
sequencing</a>), these CSV files will always be created, even if <code class="docutils literal notranslate"><span class="pre">to_csv</span></code> is <code class="docutils literal notranslate"><span class="pre">False</span></code> and the <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.align_and_parse">align_and_parse</a> function is returning <code class="docutils literal notranslate"><span class="pre">aligned</span></code> and <code class="docutils literal notranslate"><span class="pre">filtered</span></code> as dictionaries of data frames in its final output.</p>
<p>Here we only have one PacBio run, but in practice we will often have multiple. As such, our example shows how to concatenate the <code class="docutils literal notranslate"><span class="pre">readstats</span></code>, <code class="docutils literal notranslate"><span class="pre">aligned</span></code>, and <code class="docutils literal notranslate"><span class="pre">filtered</span></code> data frames for each PacBio run and then look at the data frames together:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">readstats</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">aligned</span> <span class="o">=</span> <span class="p">{</span><span class="n">targetname</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">}</span>
<span class="n">filtered</span> <span class="o">=</span> <span class="p">{</span><span class="n">targetname</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">}</span>

<span class="k">for</span> <span class="n">run</span> <span class="ow">in</span> <span class="n">pacbio_runs</span><span class="o">.</span><span class="n">itertuples</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Parsing PacBio run </span><span class="si">{</span><span class="n">run</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">run_readstats</span><span class="p">,</span> <span class="n">run_aligned</span><span class="p">,</span> <span class="n">run_filtered</span> <span class="o">=</span> <span class="n">targets</span><span class="o">.</span><span class="n">parse_alignment</span><span class="p">(</span>
        <span class="n">run</span><span class="o">.</span><span class="n">alignments</span><span class="p">,</span> <span class="n">filtered_cs</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>

    <span class="c1"># when concatenating add the run name to keep track of runs for results</span>
    <span class="n">readstats</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">run_readstats</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">run_name</span><span class="o">=</span><span class="n">run</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
        <span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">run_aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">run_name</span><span class="o">=</span><span class="n">run</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>
        <span class="n">filtered</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">run_filtered</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">run_name</span><span class="o">=</span><span class="n">run</span><span class="o">.</span><span class="n">name</span><span class="p">))</span>

<span class="c1"># now concatenate the data frames for each run</span>
<span class="n">readstats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">readstats</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
    <span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">filtered</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="n">filtered</span><span class="p">[</span><span class="n">targetname</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Parsing PacBio run lasv_pilot
</pre></div></div>
</div>
<p>First lets look at the read stats:</p>
<p>From the known composition of the library, there should be more <code class="docutils literal notranslate"><span class="pre">LASV_Josiah_OPT</span></code> reads than <code class="docutils literal notranslate"><span class="pre">LASV_Josiah_WT</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">readstats</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>category</th>
      <th>count</th>
      <th>run_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>filtered LASV_Josiah_WT</td>
      <td>23</td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>1</th>
      <td>aligned LASV_Josiah_WT</td>
      <td>84</td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>2</th>
      <td>filtered LASV_Josiah_OPT</td>
      <td>32</td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>3</th>
      <td>aligned LASV_Josiah_OPT</td>
      <td>111</td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>4</th>
      <td>unmapped</td>
      <td>0</td>
      <td>lasv_pilot</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">readstats</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;category&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">))</span>
    <span class="o">+</span> <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~ run_name&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
        <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span> <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">1.5</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">pacbio_runs</span><span class="p">),</span> <span class="mf">2.5</span><span class="p">)</span>
    <span class="p">)</span>
<span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Now look at the information on the filtered reads. This is a bigger data frame, so we just look at the first few lines for the first target (of which there is only one anyway):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">filtered</span><span class="p">[</span><span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>filter_reason</th>
      <th>filter_cs</th>
      <th>run_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194989/ccs</td>
      <td>termini5 clip5</td>
      <td>*nc*na*nc:19*nc:113</td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4260241/ccs</td>
      <td>index clip5</td>
      <td>*ng*na*nc*na*nc</td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4260334/ccs</td>
      <td>query_clip5</td>
      <td>None</td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>3</th>
      <td>m54228_190605_190010/4391712/ccs</td>
      <td>termini5 clip5</td>
      <td></td>
      <td>lasv_pilot</td>
    </tr>
    <tr>
      <th>4</th>
      <td>m54228_190605_190010/4456843/ccs</td>
      <td>termini3 clip3</td>
      <td></td>
      <td>lasv_pilot</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
    <span class="n">target_filtered</span> <span class="o">=</span> <span class="n">filtered</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span>
    <span class="n">nreasons</span> <span class="o">=</span> <span class="n">target_filtered</span><span class="p">[</span><span class="s2">&quot;filter_reason&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
    <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span><span class="n">target_filtered</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;filter_reason&quot;</span><span class="p">))</span>
        <span class="o">+</span> <span class="n">geom_bar</span><span class="p">()</span>
        <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~ run_name&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="n">targetname</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
            <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
            <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">0.3</span> <span class="o">*</span> <span class="n">nreasons</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">pacbio_runs</span><span class="p">),</span> <span class="mf">2.5</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
<section id="Error-filtering">
<h2>Error filtering<a class="headerlink" href="#Error-filtering" title="Link to this heading">¶</a></h2>
<p>Before looking at the information for the validly aligned (not filtered) reads, it is important to get a sense of the error rate for these sequencing reads.</p>
<p>These reads do not have random barcodes on the initial viral entry protein plasmids, but we can use the <code class="docutils literal notranslate"><span class="pre">gene_accuracy</span></code> information output from constructing the <code class="docutils literal notranslate"><span class="pre">ccs</span></code>s to examine accuracy.</p>
<p>We will do this using a similar method to that implemented in the <a class="reference external" href="https://jbloomlab.github.io/alignparse/recA_DMS.html">RecA DMS libraries</a> example notebook. However, here we will plot the graphs for each target.</p>
<p>We anticipate excluding all CCSs for which the error rate for either the gene or barcode is <span class="math notranslate nohighlight">\(&gt;10^{-4}\)</span>. We specify this cutoff below.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">error_rate_floor</span> <span class="o">=</span> <span class="mf">1e-7</span>  <span class="c1"># error rates &lt; this set to this</span>
<span class="n">error_cutoff</span> <span class="o">=</span> <span class="mf">1e-4</span>

<span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
    <span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span> <span class="o">=</span> <span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">gene_error</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">numpy</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene_accuracy&quot;</span><span class="p">],</span> <span class="n">error_rate_floor</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span>
            <span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span>
                <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;run_name&quot;</span><span class="p">],</span>
                <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene_error&quot;</span><span class="p">],</span>
                <span class="n">var_name</span><span class="o">=</span><span class="s2">&quot;feature_type&quot;</span><span class="p">,</span>
                <span class="n">value_name</span><span class="o">=</span><span class="s2">&quot;error rate&quot;</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;error rate&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">geom_histogram</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="o">=</span><span class="n">error_cutoff</span><span class="p">,</span> <span class="n">linetype</span><span class="o">=</span><span class="s2">&quot;dashed&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="o">+</span> <span class="n">facet_grid</span><span class="p">(</span><span class="s2">&quot;~ feature_type&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
        <span class="o">+</span> <span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;number of CCSs&quot;</span><span class="p">),</span> <span class="n">title</span><span class="o">=</span><span class="p">(</span><span class="n">targetname</span><span class="p">))</span>
        <span class="o">+</span> <span class="n">scale_x_log10</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Next, we store all reads with an error rate <span class="math notranslate nohighlight">\(&lt;10^{-4}\)</span> in new data frames for retained sequences.</p>
<p>We will use these retained sequences for further analyses.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">retained</span> <span class="o">=</span> <span class="p">{</span><span class="n">targetname</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">}</span>
<span class="k">for</span> <span class="n">targetname</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
    <span class="n">target_retained</span> <span class="o">=</span> <span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">][</span>
        <span class="n">aligned</span><span class="p">[</span><span class="n">targetname</span><span class="p">][</span><span class="s2">&quot;gene_error&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">error_cutoff</span>
    <span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">retained</span><span class="p">[</span><span class="n">targetname</span><span class="p">]</span> <span class="o">=</span> <span class="n">target_retained</span>
</pre></div>
</div>
</div>
</section>
<section id="Data-analysis">
<h2>Data analysis<a class="headerlink" href="#Data-analysis" title="Link to this heading">¶</a></h2>
<p>Now we can examine our data in more detail.</p>
<p>We will only display a few columns so the example renders well in the documentation. First, let’s look at the <code class="docutils literal notranslate"><span class="pre">query_name</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_mutations</span></code>, and <code class="docutils literal notranslate"><span class="pre">index_sequence</span></code> columns for the first few entries in the <code class="docutils literal notranslate"><span class="pre">retained</span></code> data frame for each target:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_mutations&quot;</span><span class="p">,</span> <span class="s2">&quot;index_sequence&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">target_name</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">target_name</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">retained</span><span class="p">[</span><span class="n">target_name</span><span class="p">][</span><span class="n">display_columns</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_WT
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>gene_mutations</th>
      <th>index_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194436/ccs</td>
      <td></td>
      <td>GGTATG</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4194472/ccs</td>
      <td>T572C</td>
      <td>GGTATG</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4194509/ccs</td>
      <td></td>
      <td>GGTATG</td>
    </tr>
    <tr>
      <th>3</th>
      <td>m54228_190605_190010/4194553/ccs</td>
      <td></td>
      <td>GGTATG</td>
    </tr>
    <tr>
      <th>4</th>
      <td>m54228_190605_190010/4194576/ccs</td>
      <td></td>
      <td>AGACAC</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_OPT
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>gene_mutations</th>
      <th>index_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194382/ccs</td>
      <td></td>
      <td>GAGACG</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4194390/ccs</td>
      <td></td>
      <td>ACGACC</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4194399/ccs</td>
      <td></td>
      <td>ACGACC</td>
    </tr>
    <tr>
      <th>3</th>
      <td>m54228_190605_190010/4194439/ccs</td>
      <td></td>
      <td>ACGACC</td>
    </tr>
    <tr>
      <th>4</th>
      <td>m54228_190605_190010/4194445/ccs</td>
      <td></td>
      <td>CTTCAC</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>As seen by the <code class="docutils literal notranslate"><span class="pre">index_sequence</span></code> columns, each target here has 2 or 3 different indices that map to it. These indicies indicate different samples.</p>
<p>We can then split these data frames into sample-specific data frames based on the known starting index sequences for each target. We will store these data frames in a new dictionary, keyed by target and index.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">indices</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;LASV_Josiah_WT&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;AGACAC&quot;</span><span class="p">,</span> <span class="s2">&quot;GGTATG&quot;</span><span class="p">],</span>
    <span class="s2">&quot;LASV_Josiah_OPT&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ACGACC&quot;</span><span class="p">,</span> <span class="s2">&quot;CTTCAC&quot;</span><span class="p">,</span> <span class="s2">&quot;GAGACG&quot;</span><span class="p">],</span>
<span class="p">}</span>
<span class="n">target_idx_retained</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">index_counts</span> <span class="o">=</span> <span class="p">{</span><span class="n">target</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">}</span>
<span class="n">index_counts_dfs</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">target_name</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">[</span><span class="n">target_name</span><span class="p">]:</span>
        <span class="n">target_idx_retained</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">retained</span><span class="p">[</span><span class="n">target_name</span><span class="p">][</span>
            <span class="n">retained</span><span class="p">[</span><span class="n">target_name</span><span class="p">][</span><span class="s2">&quot;index_sequence&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">index</span>
        <span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">index_counts</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="p">(</span><span class="n">index</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">target_idx_retained</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]))</span>
        <span class="p">)</span>
    <span class="n">index_counts</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="p">(</span>
            <span class="s2">&quot;invalid&quot;</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">retained</span><span class="p">[</span><span class="n">target_name</span><span class="p">])</span>
                <span class="o">-</span> <span class="nb">sum</span><span class="p">(</span><span class="n">idx_tup</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">idx_tup</span> <span class="ow">in</span> <span class="n">index_counts</span><span class="p">[</span><span class="n">target_name</span><span class="p">])</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">index_counts_dfs</span><span class="p">[</span><span class="n">target_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">index_counts</span><span class="p">[</span><span class="n">target_name</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">]</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>In the process of making these separate data frames, we also kept track of how many reads for each target map to each index or don’t map to an index (so have an ‘invalid’ index) and can now plot these counts. As seen below, only one sequence in this example has an invalid index and it maps to the <code class="docutils literal notranslate"><span class="pre">LASV_Josiah_OPT</span></code> target.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="k">for</span> <span class="n">target_name</span> <span class="ow">in</span> <span class="n">targets</span><span class="o">.</span><span class="n">target_names</span><span class="p">:</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">index_counts_dfs</span><span class="p">[</span><span class="n">target_name</span><span class="p">]</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;target&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">target_name</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="n">id_order</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;invalid&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">][:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
    <span class="n">df</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Categorical</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">],</span> <span class="n">categories</span><span class="o">=</span><span class="n">id_order</span><span class="p">,</span> <span class="n">ordered</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">index_count_plot</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;target&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;count&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">))</span>
        <span class="o">+</span> <span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">,</span> <span class="n">position</span><span class="o">=</span><span class="s2">&quot;stack&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
            <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">vjust</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">hjust</span><span class="o">=</span><span class="mf">0.5</span><span class="p">),</span> <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="o">+</span> <span class="n">ylab</span><span class="p">(</span><span class="s2">&quot;Reads&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">xlab</span><span class="p">(</span><span class="s2">&quot;Target&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">ggtitle</span><span class="p">(</span><span class="s2">&quot;Reads per Sample Index&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">_</span> <span class="o">=</span> <span class="n">index_count_plot</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>The two target-specific data frames are now five index-specific data frames. Again we will display the <code class="docutils literal notranslate"><span class="pre">query_name</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_mutations</span></code>, and <code class="docutils literal notranslate"><span class="pre">index_sequence</span></code> columns.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">target_idx</span> <span class="ow">in</span> <span class="n">target_idx_retained</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">target_idx</span><span class="p">)</span>
    <span class="n">display</span><span class="p">(</span><span class="n">target_idx_retained</span><span class="p">[</span><span class="n">target_idx</span><span class="p">][</span><span class="n">display_columns</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_WT_AGACAC
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>gene_mutations</th>
      <th>index_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194576/ccs</td>
      <td></td>
      <td>AGACAC</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4194612/ccs</td>
      <td></td>
      <td>AGACAC</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4194613/ccs</td>
      <td></td>
      <td>AGACAC</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_WT_GGTATG
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>gene_mutations</th>
      <th>index_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194436/ccs</td>
      <td></td>
      <td>GGTATG</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4194472/ccs</td>
      <td>T572C</td>
      <td>GGTATG</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4194509/ccs</td>
      <td></td>
      <td>GGTATG</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_OPT_ACGACC
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>gene_mutations</th>
      <th>index_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194390/ccs</td>
      <td></td>
      <td>ACGACC</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4194399/ccs</td>
      <td></td>
      <td>ACGACC</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4194439/ccs</td>
      <td></td>
      <td>ACGACC</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_OPT_CTTCAC
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>gene_mutations</th>
      <th>index_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194445/ccs</td>
      <td></td>
      <td>CTTCAC</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4194501/ccs</td>
      <td></td>
      <td>CTTCAC</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4194549/ccs</td>
      <td></td>
      <td>CTTCAC</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
LASV_Josiah_OPT_GAGACG
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>query_name</th>
      <th>gene_mutations</th>
      <th>index_sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>m54228_190605_190010/4194382/ccs</td>
      <td></td>
      <td>GAGACG</td>
    </tr>
    <tr>
      <th>1</th>
      <td>m54228_190605_190010/4194467/ccs</td>
      <td></td>
      <td>GAGACG</td>
    </tr>
    <tr>
      <th>2</th>
      <td>m54228_190605_190010/4194491/ccs</td>
      <td></td>
      <td>GAGACG</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We can add mutation info for the genes in these targets using <a class="reference external" href="https://jbloomlab.github.io/alignparse/alignparse.consensus.html#alignparse.consensus.add_mut_info_cols">alignparse.consensus.add_mut_info_cols</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">target_idx</span> <span class="ow">in</span> <span class="n">target_idx_retained</span><span class="p">:</span>
    <span class="n">target_idx_retained</span><span class="p">[</span><span class="n">target_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">alignparse</span><span class="o">.</span><span class="n">consensus</span><span class="o">.</span><span class="n">add_mut_info_cols</span><span class="p">(</span>
        <span class="n">target_idx_retained</span><span class="p">[</span><span class="n">target_idx</span><span class="p">],</span>
        <span class="n">mutation_col</span><span class="o">=</span><span class="s2">&quot;gene_mutations&quot;</span><span class="p">,</span>
        <span class="n">n_sub_col</span><span class="o">=</span><span class="s2">&quot;n_gene_subs&quot;</span><span class="p">,</span>
        <span class="n">n_indel_col</span><span class="o">=</span><span class="s2">&quot;n_gene_indels&quot;</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>We can then plot the number of reads for each sample that has each number of substitutions or indels to get a sense of how many mutations are in these reads, if some smaples have more mutations than others, and if substitutions or indels are more prevalent in these sequences.</p>
<p>With such a small sample snippet of the data, it is difficult to make any conclusions, but with a full dataset, this can provide important information about the mutational processes at work for each sample.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">mut_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;n_gene_subs&quot;</span><span class="p">,</span> <span class="s2">&quot;n_gene_indels&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">target_idx</span> <span class="ow">in</span> <span class="n">target_idx_retained</span><span class="p">:</span>
    <span class="n">target_idx_plot_muts</span> <span class="o">=</span> <span class="n">target_idx_retained</span><span class="p">[</span><span class="n">target_idx</span><span class="p">][</span><span class="n">mut_cols</span><span class="p">]</span><span class="o">.</span><span class="n">melt</span><span class="p">()</span>
    <span class="n">mut_counts_plot</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">ggplot</span><span class="p">(</span><span class="n">target_idx_plot_muts</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s2">&quot;value&quot;</span><span class="p">))</span>
        <span class="o">+</span> <span class="n">geom_bar</span><span class="p">()</span>
        <span class="o">+</span> <span class="n">facet_wrap</span><span class="p">(</span><span class="s2">&quot;~ variable&quot;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">xlim</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">target_idx</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="o">+</span> <span class="n">theme</span><span class="p">(</span>
            <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
            <span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">mut_counts_plot</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo" />
    
    <h1 class="logo logo-name">alignparse</h1>
    
  </a>
</p>



<p class="blurb">Align sequences and then parse features.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=alignparse&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">alignparse documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="recA_DMS.html">RecA deep mutational scanning libraries</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Lassa virus glycoprotein PacBio sequencing</a></li>
<li class="toctree-l2"><a class="reference internal" href="flu_virus_seq_example.html">Single-cell virus sequencing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="alignparse.html">alignparse package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">Package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="examples.html">Examples</a><ul>
      <li>Previous: <a href="recA_DMS.html" title="previous chapter">RecA deep mutational scanning libraries</a></li>
      <li>Next: <a href="flu_virus_seq_example.html" title="next chapter">Single-cell virus sequencing</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2019--2025.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/lasv_pilot.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/alignparse" class="github">
        <img src="_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>