{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single-cell virus sequencing\n",
    "This example shows how to use [alignparse](https://jbloomlab.github.io/alignparse/) to process full-length PacBio CCSs of viral cDNAs generated in single-cell virus sequencing of influenza-infected cells.\n",
    "Specifically, it processes a snippet of data taking from [Russell et al, 2019](https://jvi.asm.org/content/93/14/e00500-19), which was generated by PacBio sequencing of viral cDNAs with unique molecular identifiers (UMIS) and cell barcodes added using a 10X Chromium (v2 reagents).\n",
    "This notebook aligns these CCSs to the influenza transcripts and parses the barcodes / UMIs and viral mutations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up for analysis\n",
    "Import necessary Python modules.\n",
    "We use [alignparse](https://jbloomlab.github.io/alignparse/) for most of the operations, [plotnine](https://plotnine.readthedocs.io) for ggplot2-like plotting:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-05-23T18:07:53.240786Z",
     "iopub.status.busy": "2024-05-23T18:07:53.240063Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import warnings\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from plotnine import *\n",
    "\n",
    "import alignparse.ccs\n",
    "import alignparse.consensus\n",
    "import alignparse.minimap2\n",
    "import alignparse.targets\n",
    "from alignparse.constants import CBPALETTE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Suppress warnings that clutter output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Directory for output:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = \"./output_files/\"\n",
    "os.makedirs(outdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Target amplicons\n",
    "The alignment targets consist of PCR amplicons covering the influenza virus mRNAs (WSN strain) with a 5' termini corresponding to the primer binding site, the 3' polyA tail used for reverse transcription, the UMI and cell barcode, and the common 3' termini.\n",
    "There are also variant tags distinguishing the synonymously tagged viral variants (see [Russell et al, 2019](https://jvi.asm.org/content/93/14/e00500-19)).\n",
    "\n",
    "First, let's look at the Genbank file holding the targets.\n",
    "The targets are defined in [Genbank Flat File format](https://www.ncbi.nlm.nih.gov/genbank/samplerecord/)\n",
    "Below we show the first 85 lines (the actual file is quite large as it defined all 10 influenza mRNAs):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "targetfile = \"input_files/flu_WSN_amplicons.gb\"\n",
    "\n",
    "nlines_to_show = 85\n",
    "with open(targetfile) as f:\n",
    "    print(\"\".join(next(f) for _ in range(nlines_to_show)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also have the YAML file specifying how to parse the features in the alignments.\n",
    "Note how this file uses the YAML syntax for defaults to avoid repeating the same specifications for each target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_parse_specs_file = \"input_files/flu_WSN_feature_parse_specs.yaml\"\n",
    "with open(feature_parse_specs_file) as f:\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now read the targets into an [alignparse.targets.Targets](https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets) object with the feature-parsing specs (ignoring the keys that define thet defaults):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "targets = alignparse.targets.Targets(\n",
    "    seqsfile=targetfile,\n",
    "    feature_parse_specs=feature_parse_specs_file,\n",
    "    ignore_feature_parse_specs_keys=[\"default_2tags\", \"default_4tags\"],\n",
    "    allow_extra_features=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the [Targets](https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = targets.plot(ax_width=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Several things of note about how these targets are defined.\n",
    "These all relate to the fact that the most difficult region to parse is the polyA tail and the UMI / cell barcode--the reason being that the polyA tail is not sequenced very accurately (it is a homopolymer) plus appears to be somewhat impure in the actual primers, so is difficult to align correctly.\n",
    "This is compounded by the fact that we don't know the actual UMI sequence and so don't want indels of the polyA tail to \"bleed\" into the UMI or gene alignment.\n",
    "Therefore:\n",
    "\n",
    " - We define a separate feature for the final nucleotides (last 3) of the mRNA upstream of the sequenced mRNA, so polyA indels don't get assigned to the sequenced mRNA too often.\n",
    " - We define the polyA to be just 28 nucleotides in the alignment even though the length in the primer is 30, so that deletions don't affect the UMI.\n",
    " \n",
    "These empirically help with the alilgnments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PacBio CCSs\n",
    "Now let's look at the PacBio CCSs.\n",
    "First, define a data frame with the information on the PacBio runs (here we just have one):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pacbio_runs = pd.DataFrame(\n",
    "    {\"name\": [\"flu_WSN\"], \"fastq\": [\"input_files/flu_WSN_ccs.fastq\"]}\n",
    ")\n",
    "\n",
    "pacbio_runs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create an [alignparse.ccs.Summaries](https://jbloomlab.github.io/alignparse/alignparse.ccs.html#alignparse.ccs.Summaries) object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccs_summaries = alignparse.ccs.Summaries(pacbio_runs, report_col=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Statistics on the CCSs (length, number of subread passes, quality):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NBVAL_IGNORE_OUTPUT\n",
    "for stat in [\"length\", \"passes\", \"accuracy\"]:\n",
    "    if ccs_summaries.has_stat(stat):\n",
    "        p = ccs_summaries.plot_ccs_stats(stat)\n",
    "        p = p + theme(panel_grid_major_x=element_blank())  # no vertical grid lines\n",
    "        _ = p.draw()\n",
    "    else:\n",
    "        print(f\"No {stat} statistics available.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align and parse\n",
    "First, we create an [alignparse.minimap2.Mapper](https://jbloomlab.github.io/alignparse/alignparse.minimap2.html#alignparse.minimap2.Mapper) to run [minimap2](https://github.com/lh3/minimap2), which is used for the alignments. We use [minimap2](https://github.com/lh3/minimap2) options that are tailored for viruses with potentially large internal deletions (which are handled like spliced introns); these options are specified by [alignparse.minimap2.OPTIONS_VIRUS_W_DEL](https://jbloomlab.github.io/alignparse/alignparse.minimap2.html#alignparse.minimap2.OPTIONS_VIRUS_W_DEL):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NBVAL_IGNORE_OUTPUT\n",
    "\n",
    "mapper = alignparse.minimap2.Mapper(alignparse.minimap2.OPTIONS_VIRUS_W_DEL)\n",
    "\n",
    "print(\n",
    "    f\"Using `minimap2` {mapper.version} with these options:\\n\"\n",
    "    + \" \".join(mapper.options)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now use [Targets.align_and_parse](https://jbloomlab.github.io/alignparse/alignparse.targets.html#alignparse.targets.Targets.align_and_parse) to align and parse the CCSs.\n",
    "First, create the output directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "align_and_parse_outdir = os.path.join(outdir, \"flu_WSN_align_and_parse\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now do the alignments and parsing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "readstats, aligned, filtered = targets.align_and_parse(\n",
    "    df=pacbio_runs,\n",
    "    mapper=mapper,\n",
    "    outdir=align_and_parse_outdir,\n",
    "    name_col=\"name\",\n",
    "    queryfile_col=\"fastq\",\n",
    "    overwrite=True,  # overwrite any existing output\n",
    "    ncpus=-1,  # use all available CPUs\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, let's look at the read alignment stats:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "readstats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot these stats:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NBVAL_IGNORE_OUTPUT\n",
    "p = (\n",
    "    ggplot(\n",
    "        readstats.assign(\n",
    "            category=lambda x: pd.Categorical(\n",
    "                x[\"category\"], x[\"category\"].unique(), ordered=True\n",
    "            ),\n",
    "            is_aligned=lambda x: x[\"category\"].str.contains(\"aligned\"),\n",
    "        ),\n",
    "        aes(\"category\", \"count\", fill=\"is_aligned\"),\n",
    "    )\n",
    "    + geom_bar(stat=\"identity\")\n",
    "    + facet_wrap(\"~ name\", nrow=1)\n",
    "    + theme(\n",
    "        axis_text_x=element_text(angle=90),\n",
    "        figure_size=(0.3 * len(readstats), 2.5),\n",
    "        panel_grid_major_x=element_blank(),  # no vertical grid lines\n",
    "    )\n",
    "    + scale_fill_manual(values=CBPALETTE)\n",
    ")\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's look at the reason that reads that were filtered failed to align.\n",
    "Recall that `filtered` holds a data frame for each target:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for target in targets.target_names[:1]:\n",
    "    print(f\"First few lines of `filtered` for {target}:\")\n",
    "    display(filtered[target].head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concatenate the information for all of the targets and plot reasons why mapped reads were filtered:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NBVAL_IGNORE_OUTPUT\n",
    "p = (\n",
    "    ggplot(\n",
    "        pd.concat([df.assign(gene=gene) for gene, df in filtered.items()]).assign(\n",
    "            gene=lambda x: pd.Categorical(x[\"gene\"], x[\"gene\"].unique(), ordered=True)\n",
    "        ),\n",
    "        aes(\"filter_reason\"),\n",
    "    )\n",
    "    + geom_bar()\n",
    "    + facet_wrap(\"~ gene\", ncol=5)\n",
    "    + theme(\n",
    "        axis_text_x=element_text(angle=90),\n",
    "        figure_size=(9, 4),\n",
    "        panel_grid_major_x=element_blank(),  # no vertical grid lines\n",
    "    )\n",
    ")\n",
    "\n",
    "_ = p.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's look at the first few aligned reads for the first few genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for target in targets.target_names[:4]:\n",
    "    print(f\"\\nFirst few entries in `aligned` for {target}:\")\n",
    "    display(aligned[target].head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
